---
title: "T. gonndii BMDC interaction - host analysis"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First load the required packaged into the R environment. 

```{r, load required packages, warning=FALSE}

library(Seurat)
library(dplyr)
library(Matrix)
library(gtable)
library(grid)
library(gridExtra)
library(rlang)
library(AnnotationDbi)
library(ggplot2)
library(tidyr)
library(plotly)
library(pheatmap)
library(forcats)
library(patchwork)
library(sctransform)
library(openxlsx)
suppressMessages(require(biomaRt))
library(pheatmap)
library(RCurl)
library(WGCNA)
library(ggridges)
library(clustifyr)
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)

```

```{r, define colors for paper}

grayscale = c("gray5","gray15", "gray30", "gray55", "gray70", "gray85", "gray95", "gray100")

#toxo color scale
toxo.c = c("darkgreen", "olivedrab3", "maroon4", "hotpink2")


#cluster colors
c.cols <- c("#f93768", "goldenrod", "#827717", "#2E7D32","#0097A7", "#E64A19", "blue2", "#7E57C2", "red4")

#cellclass colors
cell.cols <- c("dodgerblue2", "darkgreen", "olivedrab3", "palegoldenrod","red4", "maroon4", "hotpink2","rosybrown1")

#celltype colors 

type.cols <- c("turquoise4", "mediumorchid4")

m.cycle.c <- c("darkolivegreen", "darkorange3", "royalblue4")

```



```{r, import data from mapping}

cnt_131 <- read.table("~/t.gondi_dc/data/cnt/feat_count/count131.csv.gz", header = T,sep = "\t", row.names = 1)
head(cnt_131)
cnt_132 <- read.table("~/t.gondi_dc/data/cnt/feat_count/count132.csv.gz", header = T,sep = "\t", row.names = 1)

meta <- read.table("~/t.gondi_dc/data/meta/properanno.csv", header = T, sep = ",", row.names = 1)
head(meta)

meta2 <- read.table("~/t.gondi_dc/data/meta/latest_mapping.csv", header = T, sep = ",")

meta_131 <- meta2[grep(pattern = "131_", meta$cellname),]
meta_132 <- meta2[grep(pattern = "132_", meta$cellname),]

```

```{r, format cnt_marix}

#plate1
cnt_131 <- cnt_131[,6:389]
head(cnt_131)

names <- sapply(colnames(cnt_131),
               function(x) {strsplit(x,'_')[[1]][2]})
colnames(cnt_131) <- names
colnames(cnt_131) <- gsub("X", "", colnames(cnt_131))
colnames(cnt_131) <- gsub("131.", "", colnames(cnt_131))
colnames(cnt_131) <- paste0("131", sep = "_", colnames(cnt_131))
head(cnt_131)

#plate2
cnt_132 <- cnt_132[,6:389]
head(cnt_132)

names <- sapply(colnames(cnt_132),
               function(x) {strsplit(x,'_')[[1]][1]})
colnames(cnt_132) <- names
colnames(cnt_132) <- gsub("X", "", colnames(cnt_132))
colnames(cnt_132) <- gsub("132.", "", colnames(cnt_132))
colnames(cnt_132) <- paste0("132", sep = "_", colnames(cnt_132))
head(cnt_132)

dc.data <- cbind(cnt_131, cnt_132)

```



Import the count matrix of the single cell sequencing experiment and subset it, so it only contains the counts for the mouse genome

```{r, import expression data and subset it}

#import expression matrix first 

dc.data <- read.csv("~/t.gondi_dc/data/cnt/cnt_mus_toxo_raw.csv", row.names = 1, header = T, check.names = FALSE)

#subset for only mouse genes: 
mus.genes <- rownames(dc.data)[grep("ENSMUSG",rownames(dc.data))]

head(mus.genes)
dc.data.mus <- dc.data[mus.genes,]

```

```{r, get count for toxo genes}

toxo.genes <- rownames(dc.data)[grep("TGME49",rownames(dc.data))]
dc.data.toxo <- dc.data[toxo.genes,]

dc.data.toxo[1:5,1:5]
nUMI_toxo <-  as.data.frame(colSums(dc.data.toxo))

se.dc.mus <- AddMetaData(se.dc.mus, metadata = nUMI_toxo, col.name = "nUMI_toxo")


```

Convert ensembl-Ids to gene symbols in the matrix

```{r make gene symbol annotation file}
#Create a vector from the column names 
gn <- as.vector(rownames(dc.data.mus))
gn <- sapply(gn, function(x) {strsplit(x,'\\.')[[1]][1]} )
gn <- as.vector(unlist(gn))
rownames(dc.data.mus) <- gn

#Create an annotation file 

#mouse
mus_ensembl <- AnnotationDbi::mapIds(org.Mm.eg.db::org.Mm.eg.db,
                                      keys=gn,
                                      column="SYMBOL",
                                      keytype="ENSEMBL",
                                      multiVals="first")

#Select only the annotations that are not NA 
keep <- (!(is.na(mus_ensembl)))
genes <- mus_ensembl[keep]

#get the names only
extract <- names(mus_ensembl)[keep] 
##remove duplicates
#For ensIds
extract <- as.vector(extract[!(duplicated(extract))])
#For gene symbols
genes <- as.vector(genes[!(duplicated(extract))])

#for the expression matrix
dc.data.mus <- dc.data.mus[extract,]

#give new column names with gene symbols
rownames(dc.data.mus) <- make.unique(genes)

nmat[1:5,1:5]

```

Create the seurat object for analysis - set 2 markers for "mature" and "immature" cell types of BDMCs to get an impression of the proportions of these two types for each condition in the data-set. However I would like to use the mature/immature classification we have calculated - find a way to use this instead if comparing numbers 

```{r, Create Seuratobject, width = 10, height = 40}

meta_new <- read.table("~/t.gondi_dc/data/meta/meta_141221.csv",sep = ";", header = T, row.names = 1)

se.dc.mus <- CreateSeuratObject(nmat, min.cells = 3, min.features = 200, project = "DC-TG-FH-M", meta.data = meta_new)

se.dc.mus[["percent.mt"]] <- PercentageFeatureSet(se.dc.mus,pattern = "mt")
se.dc.mus[["percent.mature"]] <- PercentageFeatureSet(se.dc.mus, features = "mature")
se.dc.mus[["percent.immature"]] <- PercentageFeatureSet(se.dc.mus, features= "immature")


VlnPlot(se.dc.mus, features = c("nCount_RNA","nFeature_RNA", "percent.mt", "nUMI_toxo"), ncol = 2, group.by = "cell_class", cols = grayscale)

bag1.v <- VlnPlot(se.toxo, features = c("TGME49-259020"), sort = T, group.by = "cell_class", cols = toxo.c , ncol = 1)

prf.v <- VlnPlot(se.toxo, features = c("TGME49-293690"), sort = T, group.by = "cell_class", cols = c("maroon4","darkgreen","olivedrab3","hotpink2") , ncol = 1)

mic2.v <- VlnPlot(se.toxo, features = c("TGME49-201780"), sort = T, group.by = "cell_class", cols = c("hotpink2","olivedrab3","maroon4","darkgreen") , ncol = 1)

sag1.v <- VlnPlot(se.toxo, features = c("TGME49-201780"), sort = T, group.by = "cell_class", cols = c("olivedrab3","hotpink2","darkgreen","maroon4") , ncol = 1)


pdf("~/t.gondi_dc/res/figures/fig1/vln_toxo.pdf", width = 15, height = 7.5)
cowplot::plot_grid(bag1.v,prf.v,mic2.v,sag1.v)
dev.off()

pdf("~/t.gondi_dc/res/figures/fig1/vln_toxo_fraction.pdf", width = 7.5, height = 7.5)
VlnPlot(se.toxo, features = c("toxo_nUMI_p","toxo_nUMI") , group.by = "cell_class", cols = toxo.c)
dev.off()

rpl9.v <- VlnPlot(se.dc.mus, features = c("Rpl9"), sort = T,group.by = "cell_class", cols = c("dodgerblue2","maroon4","darkgreen","hotpink2","rosybrown1" ,"red4", "olivedrab3","palegoldenrod") , ncol = 1)
rpl9.v

hprt.v <- VlnPlot(se.dc.mus, features = c("Hprt"), sort = T,group.by = "cell_class", cols = c("maroon4","darkgreen","hotpink2","palegoldenrod","olivedrab3","rosybrown1" ,"red4", "dodgerblue2")  , ncol = 1)
hprt.v 

ifit1.v <- VlnPlot(se.dc.mus, features = c("Ifit1"), sort = T,group.by = "cell_class", cols = c("red4","rosybrown1","palegoldenrod","olivedrab3" , "dodgerblue2","hotpink2","maroon4","darkgreen") , ncol = 1)
ifit1.v

egr1.v <- VlnPlot(se.dc.mus, features = c("Egr1"), sort = T,group.by = "cell_class", cols = c("darkgreen","olivedrab3" ,"hotpink2","maroon4", "dodgerblue2","red4","palegoldenrod","rosybrown1")  , ncol = 1)
egr1.v 


pdf("~/t.gondi_dc/res/figures/fig1/vln_all.pdf", width = 15, height = 7.5)
cowplot::plot_grid(rpl9.v, hprt.v, ifit1.v,egr1.v, ncol = 2, nrow = 2)
dev.off()

#subset the cells with a high percentage of mitochondrial genes 
se.dc.mus <- subset(se.dc.mus, subset = percent.mt < 10)


#png(filename = "~/Dropbox/shared_ankarklev/Franzi-analysis/qc/RNA-andFeatureCount-DC.png", width = 35, height = 15, res = 300, units = "cm") #Units are pixels by default
print(VlnPlot(se.dc.mus, features = c("nCount_RNA", "nFeature_RNA"), ncol = 2, group.by = "cell_class", cols = grayscale))
#dev.off()

```

```{r, boxplot for number of toxo-transcripts / condition}

df <- data.frame(features = se.dc.mus$toxo_nUMI_p, condition = se.dc.mus$cell_class)
ggplot(df, aes(x= condition, y = features*100)) + 
  geom_boxplot()

VlnPlot(se.dc.mus, features = "toxo_nUMI", group.by = "seurat_clusters", cols = c.cols )

```


```{r, get data for cell specific transcript proportions}

#get annotation table including biotypes of genes 
genes.table.tg.mus <- read.delim("~/t.gondi_dc/data/meta/genes_annotation_TG_MUS.tab")

dc.data.p <- dc.data[intersect(rownames(dc.data), subset(genes.table.tg.mus, gene_biotype == "protein_coding")$ensembl_gene_id),]

#Toxoplasma and mouse specific genes stored in a dataframe
mouse_genes <- intersect(rownames(dc.data), subset(genes.table.tg.mus, species == "mouse")$ensembl_gene_id)

toxo_genes <- intersect(rownames(dc.data.p), subset(genes.table.tg.mus, species == "toxoplasmagondii")$ensembl_gene_id)

qc <- data.frame(nUMI = colSums(dc.data), 
                 mouse_nUMI = colSums(dc.data[mouse_genes, ]),
                 toxo_nUMI = colSums(dc.data[toxo_genes, ]),
                 nGene = apply(dc.data, 2, function(x) sum(x>2)),
                 mouse_nGene = apply(dc.data[mouse_genes, ], 2 , function(x) sum(x>2)),
                 toxo_nGene = apply(dc.data[toxo_genes, ], 2 , function(x) sum(x>2)),
                 cell_ID = factor(1:ncol(dc.data)), 
                 cell = colnames(dc.data))

qc[, c("mouse_nUMI_p", "toxo_nUMI_p")] <- qc[, c("mouse_nUMI", "toxo_nUMI")]/qc[, c("nUMI")]

qc[, c("mouse_nGene_p", "toxo_nGene_p")] <- qc[, c("mouse_nGene", "toxo_nGene")]/qc[, c("nGene")]

qc$cell_ID <- factor(qc$cell_ID, levels = order(qc$toxo_nUMI_p, decreasing = F))

options(repr.plot.width = 10, repr.plot.height = 3)
qcm <- reshape2::melt(qc, measure.vars = c("mouse_nUMI_p", "toxo_nUMI_p"))
qcm <- qcm[order(qcm$toxo_nUMI, decreasing = T), ]
qcm1 <- subset(qcm, qcm$cell %in% names(which(se.dc.mus$cell_class == "PTG_3h")))

pdf("~/t.gondi_dc/res/figures/fig1/jitter_nUMI_species.pdf", width = 5, height = 3)
    print(ggplot(qcm, aes(cell_ID, value, fill = ifelse(variable == "mouse_nUMI_p", "mouse", "toxo"), color = ifelse(variable == "mouse_nUMI_p", "mouse", "toxo_nUMI_p"))) +   geom_jitter() +
    scale_y_continuous(labels = scales::percent) + #"#4477AA", "#CC6677"
    theme(axis.text.x = element_blank()) +
    scale_fill_manual(values = c("mouse" = "#33658A", "toxo" = "#F26419")) +
    scale_color_manual(values = c("mouse" = "#33658A", "toxo" = "#F26419"), na.value = "#F26419")+ 
    ylab("amount of species specific transcripts [%]") + 
    xlab("cell") + 
    theme(legend.title = element_blank(), 
          panel.border = element_blank(),
          panel.background = element_rect("white"), 
          panel.grid.major.y = element_line(colour="lightgrey", size=0.5),
          panel.grid.major.x = element_blank()))
    dev.off()

```


```{r, normalize the data, results="hide"}

se.dc.mus <- SCTransform(se.dc.mus)

```

```{r, Run dimensionality reduction and clustering}

se.dc.mus <- RunPCA(se.dc.mus)


print(se.dc.mus[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(se.dc.mus, dims = 1:2, reduction = "pca")

DimPlot(se.dc.mus,  reduction = "pca", group.by = "cell_class", cols = "Dark2")

DimHeatmap(se.dc.mus, dims = 1:15, cells = 500, balanced = TRUE)

ElbowPlot(se.dc.mus, ndims= 20, reduction ="pca")

se.dc.mus <- FindNeighbors(se.dc.mus, dims = 1:15)
se.dc.mus <- FindClusters(se.dc.mus) #start with default resolution of 0.8
#inspect the clusters
head(Idents(se.dc.mus), 5)

se.dc.mus <- RunUMAP(se.dc.mus, dims = 1:10, reduction = "pca")
#min.dist let's me play around with the thightnes of the cells within the clusters
#the spread let's me play around with the coordinates of the distance between the clusters 
##default settings are -> min.dist: 0.3, spread: 1

#rename the clusters
new.cluster.ids <- c("M 1", "M 3", "M 4", "M 6", "M 7", "M 5", "M 2", "M 8","M 9")
names(new.cluster.ids) <- levels(se.dc.mus)
se.dc.mus <- RenameIdents(se.dc.mus, new.cluster.ids)
levels(se.dc.mus) <- as.factor(c("M 1", "M 2", "M 3","M 4", "M 5", "M 6", "M 7", "M 8", "M 9"))
levels(se.dc.mus)
       

#alternatively we can rename the seurat clusters colomun right away:
se.dc.mus$seurat_clusters <- ifelse(se.dc.mus$seurat_clusters == "1", paste("M 3"), ifelse(se.dc.mus$seurat_clusters == "0", paste("M 1"), ifelse(se.dc.mus$seurat_clusters == "6", paste("M 2"), ifelse(se.dc.mus$seurat_clusters == "4", paste("M 7"), ifelse(se.dc.mus$seurat_clusters == "2", paste("M 4"), ifelse(se.dc.mus$seurat_clusters == "5", paste("M 5"), ifelse(se.dc.mus$seurat_clusters == "3", paste("M 6"), ifelse(se.dc.mus$seurat_clusters == "7", paste("M 8"), paste("M 9")))))))))

DimPlot(se.dc.mus, group.by = "seurat_clusters", cols = c.cols)

#se.dc.mus <- readRDS("~/t.gondi_dc/res/seurat-objects/seurat_mus_only_210714.RDS")
```


Plot the Umap and visualize the clusters as well as the original cell identity

```{r, umap coordinates and plots}


umap.dc.m.clusters <- DimPlot(se.dc.mus, reduction = "umap",
                              cols = c.cols) +
                              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank())
umap.dc.m.cellclass <- DimPlot(se.dc.mus, reduction = "umap", 
                               group.by = "cell_class", 
                               cols = cell.cols)  +
                              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank(),plot.title = element_blank())


pdf("~/t.gondi_dc/res/figures/fig1/dc_umap_clusters.pdf", width = 12, height = 5) #Units are pixels by default
print(cowplot::plot_grid(umap.dc.m.clusters, umap.dc.m.cellclass))
dev.off()


pdf("~/t.gondi_dc/res/figures/fig2/cell_types_map.pdf", width = 5, height = 4) #Units are pixels by default
celltypeumap <- DimPlot(se.dc.mus, reduction = "umap", 
                               group.by = "celltype", 
                               cols = type.cols)  +
                              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank(),plot.title = element_blank())
dev.off()

pdf("~/t.gondi_dc/res/figures/fig1/dc_umap_clusters_all.pdf", width = 18, height = 5) 
print(cowplot::plot_grid(umap.dc.m.clusters, umap.dc.m.cellclass, celltypeumap, ncol = 3))
dev.off()

umap.dc.m.clusters
umap.dc.m.cellclass
```


```{r, calculate percentages of each condition within individual clusters}

# sapply function to give vector for each cluster and condition and create a dataframe with the information 
clusters <- c("M 1","M 2", "M 3" ,"M 4", "M 5", "M 6", "M 7", "M 8", "M 9")

LPS_perCluster <- sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "LPS" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

LDM_3h_perCluster <- sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "LDM_3h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

PTG_3h_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "PTG_3h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

LDM_12h_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "LDM_12h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

PTG_12h_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "PTG_12h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

ctrl_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "ctrl" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

LDM_Lys3h_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "LDM_Lys3h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

PTG_Lys3h_perCluster <-  sapply(X = clusters, function(cluster) {
  length(which(se.dc.mus@meta.data$cell_class == "PTG_Lys3h" & se.dc.mus@meta.data$seurat_clusters == cluster))
})

n1 <- clusters
identpercluster <- data.frame(n1, LDM_3h_perCluster, PTG_3h_perCluster, LDM_12h_perCluster, PTG_12h_perCluster, ctrl_perCluster, LDM_Lys3h_perCluster, PTG_Lys3h_perCluster, LPS_perCluster)
colnames(identpercluster) <- c("cluster", "LDM 3h", "PTG 3h", "LDM 12h", "PTG 12h", "control", "LDM Lysate", "PTG Lysate", "LPS")
identpercluster <- gather(identpercluster, key = "identity", value = "cellcount", -cluster)

#calculate percentages using an apply function 
percent <- function(x){
  ((x/sum(identpercluster$cellcount))*100)
}

identpercluster$percent <- identpercluster[,3]/(sum(identpercluster$cellcount))*100

#Plot results:
#png("~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/originialIDpercluster_percent.png", width = 20, height = 10, unit = "cm", res = 300)
print(ggplot(identpercluster, aes(cluster, percent, fill= identity)) +
  geom_bar(position = "stack", stat = "identity") +
  xlab("") +
  ylab ("relative proportion of cells") + 
  scale_fill_manual(values = cell.cols) + 
  theme_classic() + 
  theme(legend.title = element_blank()))
#dev.off()

```


```{r, scale the calcualted percentages}
#calculate percentages of individual clusters with normalizing to the individual cluster 

c0.p <- subset(identpercluster, cluster == "M 1")
c1.p <- subset(identpercluster, cluster == "M 2")
c2.p <- subset(identpercluster, cluster == "M 3")
c3.p <- subset(identpercluster, cluster == "M 4")
c4.p <- subset(identpercluster, cluster == "M 5")
c5.p <- subset(identpercluster, cluster == "M 6")
c6.p <- subset(identpercluster, cluster == "M 7")
c7.p <- subset(identpercluster, cluster == "M 8")
c8.p <- subset(identpercluster, cluster == "M 9")

cp.l <- c(c0.p$cellcount, c1.p$cellcount, c2.p$cellcount, c3.p$cellcount,c4.p$cellcount, c5.p$cellcount, c6.p$cellcount, c7.p$cellcount , c8.p$cellcount)


c0.p$percent.cluster <-  c0.p$cellcount/sum(c0.p$cellcount)*100
c1.p$percent.cluster <-  c1.p$cellcount/sum(c1.p$cellcount)*100
c2.p$percent.cluster <-  c2.p$cellcount/sum(c2.p$cellcount)*100
c3.p$percent.cluster <-  c3.p$cellcount/sum(c3.p$cellcount)*100
c4.p$percent.cluster <-  c4.p$cellcount/sum(c4.p$cellcount)*100
c5.p$percent.cluster <-  c5.p$cellcount/sum(c5.p$cellcount)*100
c6.p$percent.cluster <-  c6.p$cellcount/sum(c6.p$cellcount)*100
c7.p$percent.cluster <-  c7.p$cellcount/sum(c7.p$cellcount)*100
c8.p$percent.cluster <-  c8.p$cellcount/sum(c8.p$cellcount)*100

identpercluster1 <- dplyr::bind_rows(c0.p,c1.p, c2.p,c3.p, c4.p, c5.p, c6.p, c7.p, c8.p)

pdf("~/t.gondi_dc/res/figures/supplementary/supp1.pdf", ) 
  print(ggplot(identpercluster1, aes(cluster, percent.cluster, fill= identity)) +
  geom_bar(position = "stack", stat = "identity") +
  xlab("") +
  ylab ("cells of original condition per cluster [%]") + 
  scale_fill_manual(values =  cell.cols)) + 
  theme_classic() + 
  theme(legend.title = element_blank())
dev.off()

```

Next we perform a variety of differential gene expression analysis based on multiple factors and conditions

1. Find markers based on the original cell class in the metadata, i.e. based on the experimental condition

```{r, Find markers for original identities}

se.dc.mus <- SetIdent(se.dc.mus, value = "cell_class")

dc.markers.original <- FindAllMarkers(se.dc.mus, min.pct = 0.25, logfc.threshold = 0.5)

dc.markers.original <- subset(dc.markers.original, p_val_adj < 0.05)

#write.xlsx(dc.markers.original, "~/Dropbox/shared_ankarklev/Franzi-analysis/DEG-orginal-IDs.xlsx")

#Visualize in a heatmap 

top20 <- dc.markers.original %>% group_by(cluster) %>% top_n(n=20, wt =avg_log2FC)
top.20 <- DoHeatmap(se.dc.mus, features = top20$gene, group.colors = c("dodgerblue", "blue3", "forestgreen", "olivedrab", "deeppink3", "magenta4", "seagreen3", "red4"))

top.20

```

2. Find markers for the identified clusters in the cluster analysis 

```{r, Find Markers according to cluster identity and subset them for later gprofiler analysis}

se.dc.mus <- SetIdent(se.dc.mus, value = "seurat_clusters")

#Find markers for all clusters
se.dc.markers <- FindAllMarkers(se.dc.mus, min.pct = 0.25, logfc.threshold = 0.5, only.pos = T)

#subset them for only significantly differential expressed genes 
se.dc.markers <- subset(se.dc.markers, p_val_adj < 0.05)

##visualize in a heatmap: 

top10 <- se.dc.markers %>% group_by(cluster) %>% top_n(n=10, wt =avg_log2FC)
top.10 <- DoHeatmap(se.dc.mus, features = top10$gene, group.colors = c.cols)
top.10


```

Classify the cells based on their expression profile using the CLustifyr package and the microarray data of [Helft J., 2015](https://pubmed.ncbi.nlm.nih.gov/26084029/). 



```{r, load library and format microarray reference data}

##use the Helft normalized expression from the limma analysis pipeline: data.matrix

#take the mean expression of each array and gene

data.mat1 <- data.frame(GM_DC = rowMeans(data.matrix[,1:3]), GM_Mac = rowMeans(data.matrix[,4:6]))

#translate the IDS to gene symbols

#merge ensembl name and gene ID
data.mat1 <- subset(data.mat1, rownames(data.mat1) %in% intersect(rownames(data.mat1), ann$ID))
data.mat1$ID <- rownames(data.mat1)
data.mat1 <- dplyr::left_join(data.mat1, ann, by = "ID")
#set merged new name as rownames for gene entries
dup <- which(duplicated(data.mat1$gene_symbol))
data.mat1 <- data.mat1[-dup,]
rownames(data.mat1) <- data.mat1$gene_symbol

data.mat <- data.mat1[,1:2]

```

```{r, prepare necessary data from the single cell data}

##workflow optimized for seurat objects
#run clustifyr on data

se.dc.mus <- clustifyr::clustify(input = se.dc.mus, ref_mat = data.mat, cluster_col = "seurat_clusters", obj_out = T)
#used number of genes: 774
#threshold = 0.5 

head(se.dc.mus[[]])


DimPlot(se.dc.mus, group.by = "celltype", cols = type.cols)

##make a column with cell_class and type in the object 

se.dc.mus$cell_class_type <- paste0(se.dc.mus$cell_class, "_", se.dc.mus$celltype)
se.dc.mus$cluster_type <- paste0(se.dc.mus$seurat_clusters, "_", se.dc.mus$celltype)
se.dc.mus$cluster_cell_class <- paste0(se.dc.mus$seurat_clusters, "_", se.dc.mus$cell_class)

DimPlot(se.dc.mus, group.by = "celltype", cols = type.cols)

#Visualize celltype signature in a heatmap

se.dc.mus <- SetIdent(se.dc.mus, value = "celltype" )
deg_celltype <- FindAllMarkers(se.dc.mus, logfc.threshold = 3)
deg_celltype <- subset(deg_celltype, deg_celltype$p_val_adj < 0.05)

cmat <- hmat_mousetoxo(t(se.dc.mus@assays$SCT@counts), se.dc.mus$cluster_type)

deg_celltype <- deg_celltype[!duplicated(deg_celltype$gene), ]

write.table(deg_celltype, "~/t.gondi_dc/res/supplementary_data/GM-DC_GM-Mac_markers.csv", row.names = F, sep = ",", quote = F)

pdf("~/t.gondi_dc/res/figures/fig2/hmat_celltypes.pdf")
phmat_simple(cmat, seed = unique(deg_celltype$gene))
dev.off()

```


We also would like to run a cell cycle scoring analysis of the individual cell types to see whether Toxoplasma infected cells (particularly at the 12h time-point are at a specific changes in the cell cycle based on infection)

We got cell cycle associated gene lists based on he orthology analysis of Tirosh, I. et al ;  https://www-ncbi-nlm-nih-gov.ezp.sub.su.se/pmc/articles/PMC4944528/

```{r, cell cycle scoring - prepare data}

cc_file <- getURL("https://raw.githubusercontent.com/hbc/tinyatlas/master/cell_cycle/Mus_musculus.csv") 
cell_cycle_genes <- read.csv(text = cc_file)

# Acquire the S phase genes
s_ids <- cell_cycle_genes %>%
        dplyr::filter(phase == "S") %>%
        pull("geneID")
        
# Acquire the G2M phase genes        
g2m_ids <- cell_cycle_genes %>%
        dplyr::filter(phase == "G2/M") %>%
        pull("geneID")


an <- genes.table[cell_cycle_genes$geneID,]

an <- data.frame(geneID = genes.table$ensembl_gene_id, gene_name = genes.table$external_gene_name)

cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, an, by = "geneID")

s_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "S") %>%
        pull("gene_name")
        
# Acquire the G2M phase genes        
g2m_genes <- cell_cycle_markers %>%
        dplyr::filter(phase == "G2/M") %>%
        pull("gene_name")


```

```{r, cell cycle scoring - run scoring}

se.dc.mus <- CellCycleScoring(se.dc.mus, g2m.features = g2m_genes, s.features = s_genes)
DimPlot(se.dc.mus, reduction = "umap", group.by = "Phase")

pdf("~/t.gondi_dc/res/figures/supplementary/cell_cycle_scoring.pdf",width = 17.5, height = 15, res = 600, units = "cm")
DimPlot(se.dc.mus, reduction = "umap", group.by = "Phase", cols = m.cycle.c) +
                              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank())
dev.off()

```


```{r, visualization of cell cycle regulators across clusters, fig.width= 20, fig.height= 4}

# Create a data.frame with one column for continuous values, e.g. gene expression
# and one column with a group variable, e.g. clusters

cyclins <- c("Ccna1", "Ccnb1","Ccnd1", "Ccnd2", "Ccnd3", "Ccne1", "Ccne2")
cdks <- c("Cdk2", "Cdk1", "Cdk4", "Cdk6")
cdkInh <- c("Cdkn1a","Cdkn2c", "Cdkn1b","Cdkn1c","Cdkn2a","Cdkn2b")
cdkreg <- c("Cdc25a", "Cks1b","Cks2", "Gadd45a")

#cluster = vector with cell and clusterannotation/celltype annotation or annotation of interest
#gene = vector with expression values for a gene of interest in the data
#dataset = which dataset to use (dc, dc+mac,mac)
#cols = colors used for the clusters 

CyCDK_expr <- function(object, clust, gene, cols){
  #cyclin B
gg <- data.frame(cluster = clust, val = object@assays$SCT@data[gene,])
# x should be the continuous variable and y the categorical group variable
# You can color the ridges using any variable you want
  gg <- ggplot(gg, aes(x = val, y = cluster, fill = cluster)) +
  geom_density_ridges(alpha = 0.8) +# Set transparency
  scale_fill_manual(values = cols) +
  theme_ridges() +
  labs(x = "Expression value", y = "Cluster identity", fill = "Cluster identity") + 
  ggtitle(gene)
  return(gg)
}

#cyclin B
CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = c("Ccnb1"), cols = c.cols)
#cyclin D1
CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Ccnd1",cols = c.cols)
#cyclin D2
cycp1 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Ccnd2", cols = c.cols)
#cyclin D3
CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Ccnd3", cols = c.cols)
#cyclin E1
cycp2 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Ccne1", cols = c.cols)
#cyclin E2
cycp3 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Ccne2", cols = c.cols)


pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cyclin_d2.pdf")
cycp1
dev.off()

pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cyclin_e1.pdf")
cycp2
dev.off()

pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cyclin_e2.pdf")
cycp3
dev.off()

cdks <- c("Cdk2", "Cdk1", "Cdk4", "Cdk6")
#cdk1
cdk1 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Cdk1", cols = c.cols)
#cdk2
cdk2 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Cdk2", cols = c.cols)
#cdk4
cdk3 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Cdk4", cols = c.cols)
#cdk6
cdk6 <- CyCDK_expr(object = se.dc.mus, clust = se.dc.mus$seurat_clusters, gene = "Cdk6", cols = c.cols)


pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cdk1.pdf")
cdk1
dev.off()

pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cdk2.pdf")
cdk2
dev.off()

pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cdk4.pdf")
cdk3
dev.off()

pdf("~/t.gondi_dc/res/figures/supplementary/sup3_cdk6.pdf")
cdk6
dev.off()

```


```{r, run GO term analysis with gprofiler}

library(gprofiler2)

gostres.l <- setNames(lapply(subset.markers.l, function(x){
  gost(query = x[,7], organism = "mmusculus")
}), nm = c("gostres.C0", "gostres.C1", "gostres.C2", "gostres.C3", "gostres.C4", "gostres.C5", "gostres.C6","gostres.C7","gostres.C8"))

#subset for GO:BP only
gostres.BP.l <- setNames(lapply(subset.markers.l, function(x){
  gost(query = x[,7], organism = "mmusculus", sources = "GO:BP")
}), nm = c("gostres.BP.C0", "gostres.BP.C1", "gostres.BP.C2", "gostres.BP.C3", "gostres.BP.C4", "gostres.BP.C5", "gostres.BP.C6", "gostres.BP.C7", "gostres.BP.C8"))

#subset for Kegg only

gostres.KEGG.l <- setNames(lapply(subset.markers.l, function(x){
  gost(query = x[,7], organism = "mmusculus", sources = "KEGG")
}), nm = c("gostres.KEGG.C0", "gostres.KEGG.C1", "gostres.KEGG.C2", "gostres.KEGG.C3", "gostres.KEGG.C4", "gostres.KEGG.C5", "gostres.KEGG.C6", "gostres.KEGG.C7"))

#subset for reactome only

gostres.REAC.l <- setNames(lapply(subset.markers.l, function(x){
  gost(query = x[,7], organism = "mmusculus", sources = "REAC")
}), nm = c("gostres.BP.C0", "gostres.BP.C1", "gostres.BP.C2", "gostres.BP.C3", "gostres.BP.C4", "gostres.BP.C5", "gostres.BP.C6", "gostres.BP.C7", "gostres.BP.C8"))



png("~/Desktop/Reac-C8.png", width = 30, height = 15, units = "cm", res = 300)
print(ggplot(head(gostres.REAC.l$gostres.BP.C8$result, 10), aes(reorder(term_name, -log10(p_value)), -log10(p_value))) +
  geom_bar(stat = "identity", color = "black", fill = "darkblue") +
  coord_flip() +
  #scale_fill_gradientn(name = "Enrichment Score",colours =   RColorBrewer::brewer.pal(n = 20, name = "Reds")) + 
  scale_y_continuous() +
  xlab("Biological Process")+
  ylab("Enrichment Score") +
  theme(axis.text.y = element_text(size = 12), axis.title = element_blank(), panel.background = element_blank()))
dev.off()


GO.all.1 <- cowplot::plot_grid(p.go.0,NULL,p.go.4,NULL,p.go.6, ncol= 5, nrow = 1, rel_widths = c(5,0.1,5,0.1,5))
GO.all.1

GO.all.2 <- cowplot::plot_grid(p.go.1,NULL,p.go.2, ncol= 3, nrow = 1, rel_widths = c(5,0.1,5))
GO.all.2
GO.all.3 <- cowplot::plot_grid(p.go.3,NULL,p.go.5, ncol= 3, nrow = 1, rel_widths = c(5,0.1,5))
GO.all.3
GO.all.4 <- cowplot::plot_grid(p.go.7,NULL,p.go.8, ncol= 3, nrow = 1, rel_widths = c(5,0.1,5))
GO.all.4

#Save the plots all together 
#png(filename = "~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/GO-enrichment/GO-enrichments-clusterbased-4.png", width = 60, height = 10, res = 300, units = "cm") #Units are pixels by default
print(GO.all.4)
#dev.off()

#Save the plots individually

#png(filename = "~/Desktop/PhDprojectFranziskaHildebrandt/TG-DC-Analysis/Results-200507-FiguresAndTables/GO-enrichment/GO-enrichment-C0.png", width = 30, height = 15, res = 300, units = "cm") #Units are pixels by default
print(p0)
#dev.off()

```


We then can investigate the differences in infected GM-Macs infected with PTG or LDM for 12h 

```{r, look specifically into DEG between cluster 6 (12h_PTG) and cluster 7 (12h_LDM)}

#look at differentially expressed genes between the different infection clusters 
#M7 = LDM, M6 = PTG (both at 12h)
PTG_LDM_markers <- FindMarkers(se.dc.mus, ident.1 = c("M 6") ,ident.2 = c("M 7"), logfc.threshold = 0.25)
PTG_LDM_markers <- subset(PTG_LDM_markers, PTG_LDM_markers$p_val_adj < 0.05)
PTG_LDM_markers <- PTG_LDM_markers[order(-PTG_LDM_markers$avg_log2FC),]
PTG_LDM_markers$gene <- rownames(PTG_LDM_markers)

PTG_markers <- subset(PTG_LDM_markers,PTG_LDM_markers$avg_log2FC > 0)
LDM_markers <- subset(PTG_LDM_markers,PTG_LDM_markers$avg_log2FC < 0)

write.table(PTG_LDM_markers, "~/t.gondi_dc/res/supplementary_data/PTG_LDM_markers.csv", row.names = F,quote = F,sep = ",")

#heatmap
p_all <- hmat_mousetoxo(data = t(se.dc.mus@assays$SCT@counts),
               meta = se.dc.mus$seurat_clusters)

pdf("~/t.gondi_dc/res/figures/fig3/heatmap_M2_M6_M7.pdf")
phmat_simple(hmat = p_all, seed = PTG_LDM_markers$gene,
               cond = c(2,6,7))
dev.off()

heatmap.F(p_all[,c(6,7)], colors= colorRampPalette(c("#660156","#21908C", "white", "#FFF68F", "#FFD700"))(1001))

#run gsea
original.gene.list <- PTG_LDM_markers$avg_log2FC
PTG_LDM_markers$gene <- rownames(PTG_LDM_markers)
names(original.gene.list) <- PTG_LDM_markers$gene
gene_list<-na.omit(original.gene.list)
gene_list = sort(gene_list, decreasing = TRUE)
head(gene_list)

PTG_LDM_gsea <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "SYMBOL",
             exponent = 1, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.01, 
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db, 
             pAdjustMethod = "none")

head(PTG_LDM_gsea@result)

write.table(PTG_LDM_gsea@result, file = "~/t.gondi_dc/res/supplementary_data/gene_set_enrichment_M6_M7.csv", quote = F, sep = ";", row.names = F)

#get the core enrichment genes as a vector to double check where they are in the heatmap (plot only these)
core_PTG_LDM_gsea <- PTG_LDM_gsea@result$core_enrichment
core_PTG_LDM_gsea <- unique(unlist(str_split(core_PTG_LDM_gsea, "/")))

```

Heatplot shows genes in alphabetical order but I would like to have them ordered by either association with the go-term or customized, thus I will write my own function to plot these accordingly: 

```{r, custom heatplot}

#extract GO-terms and enrichment in new dataframe

go.table <- subset(PTG_LDM_gsea@result, PTG_LDM_gsea@result$Description %in% c("negative regulation of leukocyte activation","response to interferon-beta","response to bacterium","positive regulation of hydrolase activity","negative regulation of cell motility","positive regulation of cysteine-type endopeptidase activity"))

go.table

#make data frame with 1 column as the description and one column with all the genes

nrow(go.table)

l <- lapply(1:nrow(go.table), function(x){
  df <- data.frame(genes = do.call(cbind, strsplit(go.table[x,11], "/", fixed = T)), goterm= paste(go.table[x,2]))
  })

library(data.table)

df <- rbindlist(l)

#now we can add the logFCs from our genelist

logFC <- as.data.frame(subset(PTG_LDM_gsea@geneList, names(PTG_LDM_gsea@geneList) %in%  df$genes))
logFC$genes <- rownames(logFC)
colnames(logFC) <- c("logFC", "genes")
head(logFC)

#now combine the data 

library(dplyr)

df <- left_join(df, logFC, by = "genes")
df$genes <- as.factor(df$genes)

p <- df %>%
  mutate(genes = fct_relevel(genes, rev(c(
            "Ripk2", "Cx3cl1","Fyn", "Hfe", "Lag3","Socs1","Cdh1","Ccl24", "Hip1", "Pparg","Abca1","Mndal",
            "Ikbke","Ifi203","Axl",  "Cd14", "Naip2", "Cxcl3","Acod1",  "Clec4e", "Fpr2","Clec4d", "Il1b"))))

# now we can plot the data in ggplot

pdf("~/t.gondi_dc/res/figures/fig3/heatplot_M6_M7_new.pdf")
ggplot(p, aes(genes, goterm)) +
  geom_tile(aes(fill = logFC), colour = "white") + 
  scale_fill_gradientn(colors = colorRampPalette(c("#003366","#0000E5", "white", "#FF8E00", "#FF5003"))(40)) + 
  coord_flip()+
  xlab("gene")+ 
  ylab("enriched term (BP)") +
  theme(panel.background = element_rect(fill = "white"), axis.text.x = element_text(angle = 35, hjust = 1), axis.line = element_line(size=0.2, color = "black"))
dev.off()
```

Integrate Expression data of T.gondii in the mouse data by adding the clusters identified by the Toxoplasma analysis to the mouse data. We can visualize the Toxoplasma data also easily in the mouse UMAP to get a general overview (not shown in manuscript).

```{r, add toxoclusters to metadata and highlight them in umap}

toxo_clusters <- data.frame(toxo.cluster = paste0("T", " ", se.toxo$seurat_clusters), cell = rownames(se.toxo[[]]))
rownames(toxo_clusters) <- toxo_clusters$cell
names(toxo_clusters)[1] <- "toxo_cluster"
toxo_clusters$cell <- NULL
head(toxo_clusters)

se.dc.mus <- AddMetaData(se.dc.mus, metadata = toxo_clusters, col.name = "toxoclusters")

#Visualize the t. gondii clusters in the UMAP (not in manuscript)

DimPlot(se.dc.mus, group.by = "toxoclusters", split.by = "toxoclusters")

toxo.dc.0 <- DimPlot(se.dc.mus, group.by = "toxoclusters", cols = c("darkred", "gray48", "gray48", "gray48", "gray48", "gray48", "gray48"))

toxo.dc.1 <- DimPlot(se.dc.mus, group.by = "toxoclusters", cols = c("gray48", "darkred", "gray48", "gray48", "gray48", "gray48", "gray48"))

toxo.dc.2 <- DimPlot(se.dc.mus, group.by = "toxoclusters", cols = c("gray48", "gray48","darkred" ,"gray48", "gray48", "gray48", "gray48"))
  
toxo.dc.3 <- DimPlot(se.dc.mus, group.by = "toxoclusters", cols = c("gray48",  "gray48", "gray48", "darkred","gray48", "gray48"))

toxo.dc.4 <- DimPlot(se.dc.mus, group.by = "toxoclusters", cols = c("gray48",  "gray48", "gray48","gray48","darkred"))

```

```{r, add toxo cell cycle as metadata to the mouse data}

toxo_phase <- as.data.frame(paste0(se.toxo$Phase))
rownames(toxo_phase) <- rownames(se.toxo[[]])
names(toxo_phase)[1] <- "toxo_phase"
head(toxo_phase)

se.dc.mus <- AddMetaData(se.dc.mus, metadata = toxo_phase, col.name = "toxo_phase")

DimPlot(se.dc.mus, group.by = "toxo_phase")

```


We would like to compare the immune response of Toxoplasma infected DCs to the response induced by other well-established ligands for pathogen-sensing pathways. 

1. LPS (lipopolysaccharide) > TLR4
2. PAM (Pam3CSK4) > TLR2
3. high molecular weight poly(I:C) (HMW) > TLR3/MDA-5
4. CpG DNA type B (cpGB) > TLR9 
5. Sendai virus (SeV) > RIG-I
6. depleted zymosan (zymosan) > Dectin-1
7. cyclic [G(3'-5')pA(3'-5')p] (cGAMP) > STING

```{r, make gene lists}

gene.set_LPS <- read.table("~/t.gondi_dc/res/GEO_comparison/LPS_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_LPS <- gene.set_LPS$x
gene.set_LPS <- subset(gene.set_LPS, gene.set_LPS %in% rownames(se.dc.mus))

gene.set_PAM <- read.table("~/t.gondi_dc/res/GEO_comparison/PAM_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_PAM <- gene.set_PAM$x
gene.set_PAM <- subset(gene.set_PAM, gene.set_PAM %in% rownames(se.dc.mus@assays$RNA))

gene.set_HMW <- read.table("~/t.gondi_dc/res/GEO_comparison/HMW_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_HMW <- gene.set_HMW$x
gene.set_HMW <- subset(gene.set_HMW, gene.set_HMW %in% rownames(se.dc.mus@assays$RNA))

gene.set_Sev <- read.table("~/t.gondi_dc/res/GEO_comparison/SeV_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_Sev <- gene.set_Sev$x
gene.set_Sev <- subset(gene.set_Sev, gene.set_Sev %in% rownames(se.dc.mus@assays$RNA))

gene.set_cGAMP <- read.table("~/t.gondi_dc/res/GEO_comparison/cGAMP_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_cGAMP <- gene.set_cGAMP$x
gene.set_cGAMP <- subset(gene.set_cGAMP, gene.set_cGAMP %in% rownames(se.dc.mus@assays$RNA))

gene.set_cpGB <- read.table("~/t.gondi_dc/res/GEO_comparison/cpGB_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_cpGB <- gene.set_cpGB$x
gene.set_cpGB <- subset(gene.set_cpGB, gene.set_cpGB %in% rownames(se.dc.mus@assays$RNA))

gene.set_zymosan <- read.table("~/t.gondi_dc/res/GEO_comparison/Zymosan_pos_log2FC.tsv", sep = "\t", header = T)
gene.set_zymosan <- gene.set_zymosan$x
gene.set_zymosan <- subset(gene.set_zymosan, gene.set_zymosan %in% rownames(se.dc.mus@assays$RNA))


```

```{r, extract unique genes for each condition, including infections}

Sev_unique <- setdiff(gene.set_Sev, c(gene.set_cGAMP,gene.set_cpGB, gene.set_LPS, gene.set_HMW, gene.set_zymosan, gene.set_PAM))
zymosan_unique <- setdiff(gene.set_zymosan, c(gene.set_cGAMP,gene.set_cpGB, gene.set_LPS, gene.set_HMW, gene.set_Sev, gene.set_PAM))
cGAMP_unique <- setdiff(gene.set_cGAMP, c(gene.set_zymosan,gene.set_cpGB, gene.set_LPS, gene.set_HMW, gene.set_Sev, gene.set_PAM))
LPS_unique <- setdiff(gene.set_LPS, c(gene.set_zymosan,gene.set_cpGB, gene.set_cGAMP, gene.set_HMW, gene.set_Sev, gene.set_PAM))
HMW_unique <- setdiff(gene.set_HMW, c(gene.set_zymosan,gene.set_cpGB, gene.set_cGAMP, LPS_unique, gene.set_Sev, gene.set_PAM))
Pam_unique <- setdiff(gene.set_PAM, c(gene.set_zymosan,gene.set_cpGB, gene.set_cGAMP, LPS_unique, gene.set_Sev, gene.set_HMW))
cpGB_unique <- setdiff(gene.set_cpGB, c(gene.set_zymosan,gene.set_PAM, gene.set_cGAMP, LPS_unique, gene.set_Sev, gene.set_PAM))

```

Gene set expression across conditions (averaged over cells) can be visualized in a heatmap

```{r, heatmap function}

#make annotation_col
annotation_col <- data.frame(CellType = factor(rep(c("GM_DC", "GM_Mac"), each = 8)))

colorder <- c("ctrl_GM_DC", "PTG_Lys3h_GM_DC" , "PTG_3h_GM_DC", "PTG_12h_GM_DC", "LDM_Lys3h_GM_DC", "LDM_3h_GM_DC" , "LDM_12h_GM_DC", "LPS_GM_DC", "ctrl_GM_Mac", "PTG_Lys3h_GM_Mac" , "PTG_3h_GM_Mac", "PTG_12h_GM_Mac", "LDM_Lys3h_GM_Mac", "LDM_3h_GM_Mac" , "LDM_12h_GM_Mac", "LPS_GM_Mac")

rownames(annotation_col) <- colorder
#add celltypes to the data in a complexheatmap
annotation_col$Strain <- paste(c("Control", "PTG", "PTG", "PTG", "LDM", "LDM", "LDM", "LPS", "Control", "PTG", "PTG", "PTG", "LDM", "LDM", "LDM", "LPS"))
annotation_col$Time <- as.integer(paste(c(3,3,3,12,3, 3, 12, 3, 3, 3, 3, 12, 3, 3,12, 3)))



#create dataframe with colors 

#data refers to a sc expression matrix you want to perform the analysis on 
#meta refers to the data you want to group your data by (e.g. clusters, celltypes, etc.)
#cond refers to the conditions you would like to display in the heatmap (numbers 1:16) (default is all conditions)
#seed refers to the features you would like to include (charactervector of genes to display)
#ann refers to the annotations you would like to have displayed in the heatmap  (default is all 3 annotations)

mat <- t(as.matrix(se.dc.mus@assays$SCT@data))
ann_cols <- list(CellType = c("GM_DC" = "mediumorchid4", "GM_Mac" = "turquoise4"), Strain = c("PTG" = "darkorange3", "LDM" = "blue3", "Control" = "forestgreen", "LPS" = "red3"), Time = c("gray20", "gray87"))

hmat_o <- hmat_mousetoxo(data = mat, meta = se.dc.mus$cell_class_type)

#make the list shorter and ensure the genes shown in the GO analysis are highlighted 

#for main figure: heatmap lps, heatmap cpGBs - add the remaining ones in the supplement
#pdf("~/t.gondi_dc/res/figures/fig2/heatmap_lps.pdf")
phmat_original(hmat = hmat_o,
               cond = c(1:16), 
               seed = LPS_unique)
#dev.off()

#pdf("~/t.gondi_dc/res/figures/fig2/heatmap_cpGB.pdf")
phmat_original(hmat = hmat_o,
               cond = c(1:16), 
               seed = cpGB_unique)


pdf("~/t.gondi_dc/res/figures/fig2/heatmap_zymosan.pdf")
phmat_original(hmat = hmat_o,
               cond = c(1:16), 
               seed = zymosan_unique)
dev.off()

```


```{r, compare data}

se.dc.mus <- SetIdent(se.dc.mus, value = "cell_class")
cellclass.markers <- FindAllMarkers(se.dc.mus, logfc.threshold = 0.5)
se.dc.mus <- SetIdent(se.dc.mus, value = "seurat_clusters")
#identify differential expressed genes for each original condition 
LDM_12h_markers <- subset(cellclass.markers, cellclass.markers$cluster == "LDM_12h")
LDM_12h_markers <- subset(LDM_12h_markers$gene, LDM_12h_markers$p_val_adj < 0.05)
LDM_3h_markers <- subset(cellclass.markers, cellclass.markers$cluster == "LDM_3h")
LDM_3h_markers <- subset(LDM_3h_markers$gene, LDM_3h_markers$p_val_adj < 0.05)

PTG_12h_markers <- subset(cellclass.markers, cellclass.markers$cluster == "PTG_12h")
PTG_12h_markers <- subset(PTG_12h_markers$gene,  PTG_12h_markers$p_val_adj < 0.05)
PTG_3h_markers <- subset(cellclass.markers, cellclass.markers$cluster == "PTG_3h")
PTG_3h_markers <- subset(PTG_3h_markers$gene, PTG_3h_markers$p_val_adj < 0.05)



#Look at intersection differential expressed genes 

#make list for each toxo-condition and plot them in a table 
prr_list <- list(gene.set_LPS, gene.set_PAM, gene.set_cpGB, gene.set_HMW, gene.set_cGAMP, gene.set_Sev, gene.set_zymosan)
prr_unique <- set_names(list(LPS_unique, Pam_unique, cpGB_unique, HMW_unique, cGAMP_unique, Sev_unique, zymosan_unique), nm = c("LPS", "prr", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan"))


PTG_12h_int <- set_names(sapply(prr_unique, function(x){
  intersect(PTG_12h_markers, x)
}), nm = c("LPS", "PAM", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan"))

PTG_3h_int <- set_names(sapply(prr_unique, function(x){
  intersect(PTG_3h_markers, x)
}), nm = c("LPS", "PAM", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan"))

LDM_12h_int <- set_names(sapply(prr_unique, function(x){
  intersect(LDM_12h_markers, x)
}), nm = c("LPS", "PAM", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan"))

LDM_3h_int <- set_names(sapply(prr_unique, function(x){
  intersect(LDM_3h_markers, x)
}), nm = c("LPS", "PAM", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan"))


PTG_12h_int
PTG_12h_int
PTG_3h_int
LDM_12h_int
LDM_3h_int

overlap <- data.frame(PTG_3h = sapply(PTG_3h_int, function(x){length(x)}), PTG_12h = sapply(PTG_12h_int, function(x){length(x)}), LDM_3h = sapply(LDM_3h_int, function(x){length(x)}), LDM_12h = sapply(LDM_12h_int, function(x){length(x)}) )

overlap$PRR <- rownames(overlap)
overlap <- gather(overlap, "condition", value = "nGenes", 1:4)
overlap


update_geom_defaults("point",list(size=3))

#Export figure for figure 2
ggplot(overlap, aes(PRR, nGenes, color = condition)) + 
  geom_point(position = position_dodge(width=0.3), na.rm = T) +
  scale_color_manual(values = cell.cols[c(2,3,6,7)]) + 
  theme_classic() 
  
```

```{r, create a binary heatmap to look at gene expression in each condition, fig.height= 5, fig.width=5 }

#create an empy dataframe 
prr_list <- c(gene.set_LPS, gene.set_PAM, gene.set_cpGB, gene.set_HMW, gene.set_cGAMP, gene.set_Sev, gene.set_zymosan)
prr_unique <- c(LPS_unique, Pam_unique, cpGB_unique, HMW_unique, cGAMP_unique, Sev_unique, zymosan_unique)

#intersect prr with the toxodata 
PTG_12h_int <- intersect(PTG_12h_markers, prr_unique)
PTG_3h_int <- intersect(PTG_3h_markers, prr_unique)
LDM_12h_int <- intersect(LDM_12h_markers, prr_unique)
LDM_3h_int <- intersect(LDM_3h_markers, prr_unique)

toxo_int <- union(LDM_3h_int, LDM_12h_int)
toxo_int <- union(toxo_int,PTG_3h_int)  
toxo_int <- union(toxo_int, PTG_12h_int)

genes <- union(prr_unique, toxo_int)

bin.hmat <- data.frame(matrix(ncol = 11, nrow = length(genes)))

colnames(bin.hmat) <- c("LPS", "PAM", "cpGB", "HMW", "cGAMP", "SeV", "Zymosan", "LDM_3h", "LDM_12h", "PTG_3h", "PTG_12h")
rownames(bin.hmat) <- genes

bin.hmat[1:5,1:5]

#fill empty dataframe with values of each gene iterating over rows and columns 


bin.hmat[,1] <- ifelse(rownames(bin.hmat) %in% gene.set_LPS, 1, 0)
bin.hmat[,2] <- ifelse(rownames(bin.hmat) %in% gene.set_PAM, 1, 0)
bin.hmat[,3] <- ifelse(rownames(bin.hmat) %in% gene.set_cpGB,1, 0)
bin.hmat[,4] <- ifelse(rownames(bin.hmat) %in% gene.set_HMW, 1, 0)
bin.hmat[,5] <- ifelse(rownames(bin.hmat) %in% gene.set_cGAMP, 1, 0)
bin.hmat[,6] <- ifelse(rownames(bin.hmat) %in% gene.set_Sev, 1, 0)
bin.hmat[,7] <- ifelse(rownames(bin.hmat) %in% gene.set_zymosan, 1, 0)
bin.hmat[,8] <- ifelse(rownames(bin.hmat) %in% LDM_3h_markers, 1, 0)
bin.hmat[,9] <- ifelse(rownames(bin.hmat) %in% LDM_12h_markers, 1, 0)
bin.hmat[,10] <- ifelse(rownames(bin.hmat) %in% PTG_3h_markers, 1, 0)
bin.hmat[,11] <- ifelse(rownames(bin.hmat) %in% PTG_12h_markers, 1, 0)


bin.hmat2 <- t(bin.hmat)
pheatmap(bin.hmat2, angle_col =  c(315), fontsize_col = 8, cellheight = 10, cellwidth = 10, cluster_cols = F, color = colorRampPalette(c("#440154" ,"white", "#FDE725"))(200))

#show only genes present in toxoplasma infection

tox.overlap <- which(rowSums(bin.hmat[,c(8:11)]) >= 1)

pdf("~/t.gondi_dc/res/figures/fig2/overlap_genes_hmat.pdf")
pheatmap(bin.hmat[tox.overlap,], angle_col =  c(45), fontsize_col = 8,fontsize_row = 8, cluster_rows = F,
         cellheight = 10, cellwidth = 15, cluster_cols = F, color = c("#440154","#FDE725"))
dev.off()

```


We also would like to run an co-expression network 

```{r, load libraries, load and match data}

#load package
library(WGCNA)

#get preprocessed mouse matrix
M_mouse <- as.data.frame(se.dc.mus@assays$SCT@counts)
ncol(M_mouse)
#get pre-processed toxo matrix
#se.toxo <- readRDS("~/t.gondi_dc/res/seurat-objects/toxo.object_150721.RDS")
M_toxo <- as.data.frame(se.toxo@assays$SCT@counts)
ncol(M_toxo)
#Filter columns of mouse data to only have columns containing toxo genes 

M_mouse <- dplyr::select(M_mouse, matches(colnames(M_toxo)))
ncol(M_mouse)

M_toxo <- as.matrix(dplyr::select(M_toxo, matches(colnames(M_mouse))))
ncol(M_toxo)

M_mouse <- as.matrix(M_mouse)

```

```{r, create correlation matrix}

set.seed(3)
corM = bicor(t(M_toxo),t(M_mouse))
corM[1:5,1:5]

#hist(corM)

```

So we first find the top correlated interactors for each gene, and we’re only looking for interactions between moue and toxo.
We also only save 20 interactors for each gene - it’s unlikely anyone creates KNN graphs with a higher K. If they do there are much better tools than KNN graphs.

```{r, interactor defintion}

# Is it only selecting interactors with positive correlations here? 

TopIntGenes_mouse = matrix(rownames(M_toxo)[apply(FUN=order, X=corM, MARGIN=2, decreasing=T, na.last=T)[1:20,]],nrow=20)
TopIntGenes_toxo = matrix(rownames(M_mouse)[apply(FUN=order, X=corM, MARGIN=1, decreasing=T, na.last=T)[1:20,]],nrow=20)
colnames(TopIntGenes_mouse) = rownames(M_mouse)
colnames(TopIntGenes_toxo) = rownames(M_toxo)

```


```{r, correlation values}

TopIntVal_mouse = matrix(ncol=ncol(TopIntGenes_mouse),nrow=nrow(TopIntGenes_mouse))
for(ii in 1:ncol(TopIntVal_mouse)){TopIntVal_mouse[,ii] = corM[TopIntGenes_mouse[,ii],ii]}

TopIntVal_toxo = matrix(ncol=ncol(TopIntGenes_toxo),nrow=nrow(TopIntGenes_toxo))
for(ii in 1:ncol(TopIntVal_toxo)){TopIntVal_toxo[,ii] = corM[ii,TopIntGenes_toxo[,ii]]}

#combine matrices

TopIntGenes = cbind(TopIntGenes_mouse, TopIntGenes_toxo)
#Remove Gm and mt-genes
TopIntVal = cbind(TopIntVal_mouse, TopIntVal_toxo)
colnames(TopIntVal) = colnames(TopIntGenes)


```

For the resource, we have now pre-calculated all the objects we need. The rest will be performed in real time based on user input.

The user will select the following parameters:
seed = the seed genes we build networks from
K = number of nearest neighbors
Nsteps = Number of steps we take from the seed genes.
cutoffValue = the hard cutoff value

```{r, create function for plotting, fig.width = 18, fig.height = 15}

library(igraph)


f.KNN_mousetoxo = function(seed, K=3, Nsteps=2, cutoffValue=0.65){

# We first filter down the Top Interactors tables to the chosen number of interactors.
TopIntGenes2 = TopIntGenes[1:K,]
TopIntVal2 = TopIntVal[1:K,]

# We then remove interactors with a weight below 0.65.
if(cutoffValue<0){cutoffValue = 0; print("Warning: CutoffValue forced to 0!")}
x.remove = abs(TopIntVal2) < cutoffValue
TopIntGenes3 = TopIntGenes2; TopIntGenes3[x.remove] = NA
TopIntVal3 = TopIntVal2; TopIntVal3[x.remove] = NA

# Now we pick out the nearest neighbors, Nsteps steps away.
seed2 = seed
if(Nsteps>1){for(ii in 1:(Nsteps-1)){
  seed2 = c(seed2,TopIntGenes3[,seed2])
  seed2 = unique(seed2[!is.na(seed2)])
}}
TopIntGenes4 = TopIntGenes3[,seed2]
TopIntVal4 = TopIntVal3[,seed2]

# These objects are now transformed to an edgelist:
keep.x = !is.na(as.vector(TopIntGenes4))
elist1 = rep(colnames(TopIntGenes4), each=nrow(TopIntGenes4))[keep.x]
elist2 = as.vector(TopIntGenes4)[keep.x]
weights = as.vector(TopIntVal4)[keep.x]

# There may be different fancy packages to create interactive plots, but I will use a 
# straightforward solution: igraph

gr.x = graph.data.frame(cbind(elist1,elist2), directed = T)

# You can plot this graph, but it doesn't look very good. There are many things we can do to improve it. 
V(gr.x)$label.family = "sans"
V(gr.x)$label.cex = c(0.7)
V(gr.x)$label.font = c(2)
V(gr.x)$label.color = c("#333333")
#V(gr.x)$label.dist = c(-0.5)
#V(gr.x)$label.degree = pi/6
E(gr.x)$arrow.size = 0.5
E(gr.x)$width = abs(weights)*2.5
V(gr.x)$name = make.names(V(gr.x)$name)
#V(gr.x)$color[V(gr.x)$name %in% make.names(PTG_12h_markers)] = "#66CCFF"
#V(gr.x)$color[V(gr.x)$name %in% make.names(LDM_12h_markers)] = "yellow3"
V(gr.x)$color[V(gr.x)$name %in% make.names(colnames(TopIntGenes_mouse))] = "grey"
V(gr.x)$color[V(gr.x)$name %in% make.names(PTG_LDM_markers$gene[PTG_LDM_markers$avg_log2FC > 0])] = "#ff7d00" 
V(gr.x)$color[V(gr.x)$name %in% make.names(PTG_LDM_markers$gene[PTG_LDM_markers$avg_log2FC < 0])] = "#4cc9f0"
V(gr.x)$color[V(gr.x)$name %in% make.names(colnames(TopIntGenes_toxo))] = "#f4d35e"
#V(gr.x)$frame.color[V(gr.x)$name %in% make.names(colnames(TopIntGenes_mouse))] = "#3399CC"
#V(gr.x)$frame.color[V(gr.x)$name %in% make.names(colnames(TopIntGenes_toxo))] = "yellow4"
#V(gr.x)$shape[V(gr.x)$name %in% make.names(colnames(TopIntGenes_mouse))] = "circle"
#V(gr.x)$shape[V(gr.x)$name %in% make.names(colnames(TopIntGenes_toxo))] = "circle"

# We set the vertex sizes according to (minimum) distance from seed genes
dist.x = distances(gr.x)
for(ii in 1:ncol(dist.x)){dist.x[ii,ii] = NA}
dist.x = dist.x[rownames(dist.x)%in%seed,]
dist.x2 = apply(FUN=min, X=dist.x, MARGIN=2,na.rm=T)
V(gr.x)$size = 2+3*(2^-dist.x2[V(gr.x)$name])


# Color for interactions with positive correlations is set to red. 
# Negative, if such exist, are colored blue.
E(gr.x)$color[weights>0] = "#CC3333"
E(gr.x)$color[weights<0] = "blue"

# We can also map information to colors and shape of vertices.
# There are two obvious options: 
# 1) Different shapes for mouse and toxo, node colors according to a selected cluster from the umap.
# 2) Make each node a pie chart with one piece for each cluster. Then distinction between mouse
#    and toxo will have to be made some other way.
#
# I will leave this choice to a future discussion. For now, here is a simple function to translate
# numeric values into a color scale, like a typical heatmap.

f.TranslateToColor = function(val){
  bins.x = seq(from=-max(abs(val),na.rm=T)-0.01, to=max(abs(val),na.rm=T)+0.01, length.out=101)
  cval = c()
  for(ii in 1:length(val)){cval[ii] = sum(bins.x<val[ii])}
  cvec = colorRampPalette(c("blue", "dodgerblue", "white", "orange", "red"))(100)[cval]
  if(is.matrix(val)){
    cvec = matrix(cvec,nrow=nrow(val))
    rownames(cvec) = rownames(val)
    colnames(cvec) = colnames(val)
  } else {
    names(cvec) = names(val)
  }
  return(cvec)
}

gr.x$layout <- layout_with_kk(gr.x)

par(mar=c(0,0,0,0)+.1)
plot(gr.x, asp = 0) 
return(gr.x)


}

#make graph


pdf("~/t.gondi_dc/res/figures/fig3/PTG_LDM_markers_network_0.3.pdf", width = 20, height = 20)
gr.1 <- f.KNN_mousetoxo(PTG_LDM_markers$gene, cutoffValue = 0.3, K = 3, Nsteps = 3)
dev.off()



```

Get a list of all co-expressed mouse genes and their network identity 

```{r, make table for toxo and mouse genes from interaction graph}


go.table.tg$ensembl_gene_id <- gsub("-", ".",go.table.tg$ensembl_gene_id)
#gr = graph
#ann = annotation table - default is the BioMart table used for annotations 
get.toxo.table <- function(gr, ann = go.table.tg){
ann$ensembl_gene_id <- gsub("_", ".", ann$ensembl_gene_id)
inf <- data.frame(clusters(gr)$membership)
colnames(inf) <- c("co-expression_cluster")
inf$ensembl_gene_id <- rownames(inf)
int.table <- ann[ann$ensembl_gene_id %in% rownames(inf),]
tg.int <- dplyr::left_join(int.table,inf, by = "ensembl_gene_id")
return(tg.int)
}

tab1 <- get.toxo.table(gr.1)
write.table(tab1, "~/t.gondi_dc/res/supplementary_data/interaction_table_toxo.csv", sep = ";", quote = F, row.names = F)

##Do the same for mouse genes

get.mouse.table <- function(gr, ann = go.table){
inf <- data.frame(clusters(gr)$membership)
colnames(inf) <- c("co-expression_cluster")
inf$external_gene_name <- rownames(inf)
int.table <- ann[ann$external_gene_name %in% rownames(inf),]
mouse.int <- dplyr::left_join(int.table,inf, by = "external_gene_name")
return(mouse.int)
}

tab2 <- get.mouse.table(gr.1)
tab2$description <- gsub(";", ",", tab2$description)
write.table(tab2, "~/t.gondi_dc/res/supplementary_data/interaction_table_mouse.csv", sep = ";", quote = F, row.names = F)


```


Generate subgraphs for ease of inspection of co-expression clusters of interest and run gene ontology analysis with visualisation using treeplots

```{r, run subgraph of graph in new function, fig.width= 20, fig.height= 12}

set.seed(3)
#gr = graphobject
#goi = gene of interest

subgraph_mtoxo_goi = function(gr, goi){
mem.x = clusters(gr)$membership
index.x = mem.x[V(gr)$name == goi]
gr.x2 = induced_subgraph(gr, which(mem.x==index.x))

par(mar=c(0,0,0,0)+.1)
plot(gr.x2, asp = 0) 
return(gr.x2)
}

subgraph_mtoxo_clust = function(gr, clust){
mem.x = clusters(gr)$membership
index.x = mem.x[clusters(gr)$membership == clust]
gr.x2 = induced.subgraph(gr, which(mem.x==index.x))
gr.x2$layout <- layout_with_lgl(gr.x2)
par(mar=c(0,0,0,0)+.5)
plot(gr.x2, asp = 0) 
return(gr.x2)
}


#subgraph_mtoxo_goi(gr = gr.1, goi = "TGME49.242240"), ##you can check what your favorite genes is co-expressed with and in which co-expression cluster it appears!


pdf("~/t.gondi_dc/res/figures/fig3/go_term_intclust9_2.pdf",  width = 8, height = 8)
c9.sub <- subgraph_mtoxo_clust(gr = gr.1, clust = 9) # get subgraph for cluster 9 
dev.off()
#read annotation table 
go.table.tg <- read.table("~/t.gondi_dc/data/meta/GeneByLocusTag_Summary.csv", sep = ",", header = T)
colnames(go.table.tg) <- c("ensembl_gene_id", "source_id", "description","gene_biotype", "external_gene_name")
#get table of genes 
tab9 <- get.mouse.table(c9.sub)
#run go-term analysis 
pdf("~/t.gondi_dc/res/figures/fig3/go_term_intclust9.pdf", width = 20, height = 16)
golysis.tree(tab9$external_gene_name,nterms = 5, source = "GO:BP", significant = F)
dev.off()


#check whether the correct cluster was selected
pdf("~/t.gondi_dc/res/figures/fig3/go_term_intclust1_2.pdf", width = 10, height = 10)
c1.sub <- subgraph_mtoxo_clust(gr = gr.1, clust = 1) 
dev.off()
#get table of genes 
tab1 <- get.mouse.table(c1.sub)
tab1.tg <- get.toxo.table(c1.sub)
tab1.tg$ensembl_gene_id <- gsub("\\.", "-", tab1.tg$ensembl_gene_id)
#run go term analysis
pdf("~/t.gondi_dc/res/figures/fig3/go_term_intclust1.pdf", width = 20, height = 16)
golysis.tree(tab1$external_gene_name,nterms = 5, source = "GO:BP", significant = F)
dev.off()

```
