---
title: "DC-T-combined analysis"
output: html_notebook
---



```{r, load the data}

dc.data <- read.csv("~/Desktop/PhDprojectFranziskaHildebrandt/TG-DC-Analysis/cnt_normalized 3.csv", row.names = 1, header = T, check.names = FALSE)

genes.table.tg.mus <- read.table("~/Desktop/PhDprojectFranziskaHildebrandt/TG-DC-Analysis/genes_annotation_TG_MUS.tab", stringsAsFactors = FALSE)
sum(rownames(dc.data) %in% genes.table.tg.mus$ensembl_gene_id)
dim(dc.data)

m <- match(rownames(dc.data), genes.table.tg.mus$ensembl_gene_id)

genes.table.tg.mus$unique_gene_name <- NA
genes.table.tg.mus$unique_gene_name[is.na(genes.table.tg.mus$external_gene_name)] <- make.unique(genes.table.tg.mus$ensembl_gene_id[is.na(genes.table.tg.mus$external_gene_name)])
genes.table.tg.mus$unique_gene_name[!is.na(genes.table.tg.mus$external_gene_name)] <- make.unique(genes.table.tg.mus$external_gene_name[!is.na(genes.table.tg.mus$external_gene_name)])

rownames(genes.table.tg.mus) <- genes.table.tg.mus$ensembl_gene_id

#merge ensembl name and gene ID
dc.data <- dc.data[intersect(rownames(dc.data), rownames(genes.table.tg.mus)), ]
newnames <- genes.table.tg.mus[rownames(dc.data), ]$unique_gene_name
#newnames <- apply(cbind(as.vector(genes.table.tg.mus$unique_gene_name)[m], rownames(dc.data)), 1, paste, collapse="::")

#set merged new name as rownames for gene entries

rownames(dc.data) <- newnames

meta1 <- read.table("~/Desktop/PhDprojectFranziskaHildebrandt/TG-DC-Analysis/metadata-200430.csv", sep = ",", header = T, row.names = 1)

se.dc <- CreateSeuratObject(dc.data, min.cells = 3, min.features = 200, project = "DC-TG-FH-COMBINED", meta.data = meta1)

#Create subset without LPS and Lysate

#se.dc.f <- subset(se.dc, subset = cell_class %in% c("ctrl", "LDM_12h", "LDM_3h", "PTG_12h", "PTG_3h"))

##Remove mt-genes

se.dc[["percent.mt"]] <- PercentageFeatureSet(se.dc,pattern = "mt-")
#se.dc.f[["percent.mt"]] <- PercentageFeatureSet(se.dc.f, pattern = "mt-")

grayscale <- c("gray15","gray20","gray25","gray30", "gray35", "gray40", "gray45", "gray50")

VlnPlot(se.dc, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"), ncol = 3, group.by = "cell_class", cols = grayscale)
#VlnPlot(se.dc.f, features = c("nCount_RNA", "nFeature_RNA"), ncol = 2, group.by = "cell_class", cols = grayscale)

#Export Violinplot 

#png(filename = "~/Desktop/PhDprojectFranziskaHildebrandt/TG-DC-Analysis/Results-200507-FiguresAndTables/RNA-andFeatureCount-combined.png", width = 35, height = 15, res = 300, units = "cm") #Units are pixels by default
#print(VlnPlot(se.dc, features = c("nCount_RNA", "nFeature_RNA"), ncol = 2, group.by = "cell_class", cols = grayscale))
#dev.off()


#subset the cells with a high percentage of mitochondrial genes 
se.dc <- subset(se.dc, subset = percent.mt < 10)
#only include Toxo at both timepoints against the control for both time-points 

se.dc <- subset(se.dc, subset = cell_class %in% c("ctrl", "LDM_12h", "LDM_3h", "PTG_12h", "PTG_3h"))

```

```{r, pie-chart with percentages}

LDM3hc <- subset(se.dc, subset = cell_class == "LDM_3h") 
LDM12hc <- subset(se.dc, subset = cell_class == "LDM_12h")
PTG3hc <- subset(se.dc, subset = cell_class == "PTG_3h") 
PTG12hc <- subset(se.dc, subset = cell_class == "PTG_12h")

LDM3hc.toxo <- sum(LDM3hc@meta.data$toxo_nUMI)
LDM3hc.dc <- sum(LDM3hc@meta.data$mouse_nUMI)
LDM12hc.toxo <- sum(LDM12hc@meta.data$toxo_nUMI)
LDM12hc.dc <- sum(LDM12hc@meta.data$mouse_nUMI)

PTG3hc.toxo <- sum(PTG3hc@meta.data$toxo_nUMI)
PTG3hc.dc <- sum(PTG3hc@meta.data$mouse_nUMI)
PTG12hc.toxo <- sum(PTG12hc@meta.data$toxo_nUMI)
PTG12hc.dc <- sum(PTG12hc@meta.data$mouse_nUMI)

#also do percentages for the piecharts

LDM3hc.toxo.p <- sum(LDM3hc@meta.data$toxo_nUMI_p)
LDM3hc.dc.p <- sum(LDM3hc@meta.data$mouse_nUMI_p)
LDM12hc.toxo.p <- sum(LDM12hc@meta.data$toxo_nUMI_p)
LDM12hc.dc.p <- sum(LDM12hc@meta.data$mouse_nUMI_p)

PTG3hc.toxo.p <- sum(PTG3hc@meta.data$toxo_nUMI_p)
PTG3hc.dc.p <- sum(PTG3hc@meta.data$mouse_nUMI_p)
PTG12hc.toxo.p <- sum(PTG12hc@meta.data$toxo_nUMI_p)
PTG12hc.dc.p <- sum(PTG12hc@meta.data$mouse_nUMI_p)

#make dataframe with the calculated values (1st column = names, 2nd column = values from there we can decide how to visaulize them )

condition <- c("LDM3h-toxo", "LDM3h-mouse","LDM12h-toxo","LDM12h-mouse","PTG3h-toxo","PTG3h-mouse","PTG12h-toxo","PTG12h-mouse")
value <- c(LDM3hc.toxo,LDM3hc.dc,LDM12hc.toxo,LDM12hc.dc,PTG3hc.toxo,PTG3hc.dc,PTG12hc.toxo,PTG12hc.dc)
percent <- c(LDM3hc.toxo.p,LDM3hc.dc.p,LDM12hc.toxo.p,LDM12hc.dc.p,PTG3hc.toxo.p,PTG3hc.dc.p,PTG12hc.toxo.p,PTG12hc.dc.p)

ntoxotransripts <- data.frame(condition,value,percent)
#Format data -> add column with additional information to plot 
ntoxotransripts$time <- as.factor(c(3, 3, 12, 12, 3, 3, 12,12))
ntoxotransripts$strain <- c("LDM","LDM", "LDM", "LDM", "PTG","PTG","PTG","PTG" )
ntoxotransripts$species <- c("toxoplasma","mouse", "toxoplasma","mouse", "toxoplasma","mouse","toxoplasma","mouse")
ntoxotransripts$condition <- NULL
#Plot to compare the species at all timepoints
#stack the species on top of each other and colorcode by strain 

nLDM3h <- subset(ntoxotransripts, strain == "LDM" & time == 3)
nPTG3h <- subset(ntoxotransripts, strain == "PTG" & time == 3)
nLDM12h <- subset(ntoxotransripts, strain == "LDM" & time == 12)
nPTG12h <- subset(ntoxotransripts, strain == "PTG" & time == 12)


a <- ggplot(ntoxotransripts, aes(time, value, fill= species)) +
  geom_bar(position = "stack", stat = "identity",fill = c("goldenrod","darkcyan","goldenrod", "darkcyan", "goldenrod", "darkcyan","goldenrod", "darkcyan")) +   scale_x_discrete(breaks = c(3,12)) +
  ylab("n unique transcripts") +
  xlab("time post infection [h]") +
  theme(legend.position="right") +
  facet_wrap(~strain, ncol = 2) + 
  theme_classic()

#png("~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/numberoftranscripts-tp.png", width = 30, height = 30, units = "cm", res = 300)
#print(a)
#dev.off()

#get percentages for the transcripts in a new column 
#Run new with calculating the percentages! 

b <- ggplot(nLDM3h, aes(strain, percent, fill= species)) +
  geom_bar(width = 1, stat = "identity")
b <- b+ coord_polar(theta = "y", start = 0) + 
  theme_void() +
  theme(axis.text.x=element_blank()) + 
  scale_fill_manual(values=c("darkcyan","goldenrod"))+
  ggtitle("LDM 3h p.i.")

#somehow add percentages in the plot? / axis labels in the plot? 
c <- ggplot(nPTG3h, aes(strain, percent, fill= species)) +
  geom_bar(width = 1, stat = "identity")
c <- c + coord_polar(theta = "y", start = 0) + 
  theme_void() +
  theme(axis.text.x=element_blank()) + 
  scale_fill_manual(values=c("darkcyan","goldenrod"))+
  ggtitle("PTG 3h p.i.")

d <- ggplot(nLDM12h, aes(strain, percent, fill= species)) +
  geom_bar(width = 1, stat = "identity")
d <- d + coord_polar(theta = "y", start = 0) + 
  theme_void() +
  theme(axis.text.x=element_blank()) + 
  scale_fill_manual(values=c("darkcyan","goldenrod"))+
  ggtitle("LDM 12h p.i.", )

e <- ggplot(nPTG12h, aes(strain, percent, fill= species)) +
  geom_bar(width = 1, stat = "identity")
e <- e + coord_polar(theta = "y", start = 0) + 
  theme_void() +
  theme(axis.text.x=element_blank()) + 
  scale_fill_manual(values=c("darkcyan","goldenrod"))+
  ggtitle("PTG 12h p.i.")

#png("~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/piechart-transcript-percents.png", height = 30, width = 30, units = "cm", res = 300)
print(cowplot::plot_grid(b, NULL,c, d,NULL, e, ncol = 3, nrow = 2, rel_widths = c(1,0.3,1)))
#dev.off()

```

```{r, normalize the data}

se.dc <- SCTransform(se.dc)
#se.dc.f <- SCTransform(se.dc.f)

```


```{r, run dimensionality reduction and clustering, fig.width=12, fig.height=6}

#Define colors for the umaps: 

c.cellclass <- c("#1B9E77","#D95F02", "#7570B3" ,"#E6AB02" ,"#A6761D", "#666666")

se.dc <- RunPCA(se.dc)
#se.dc.f <- RunPCA(se.dc.f)

print(se.dc[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(se.dc, dims = 1:2, reduction = "pca")

DimPlot(se.dc,  reduction = "pca", group.by = "cell_class")

DimHeatmap(se.dc, dims = 1:15, cells = 500, balanced = TRUE)

ElbowPlot(se.dc, ndims= 20, reduction ="pca")
#ElbowPlot(se.dc.f, ndims= 20, reduction ="pca")

se.dc <- FindNeighbors(se.dc, reduction = "pca", dims = 1:10)
#se.dc.f <- FindNeighbors(se.dc.f, reduction = "pca", dims = 1:10)
se.dc <- FindClusters(se.dc)
#se.dc.f <- FindClusters(se.dc.f)
#start with default resolution of 0.8
#inspect the clusters
head(Idents(se.dc), 5)

se.dc <- RunUMAP(se.dc, reduction = "pca",dims = 1:10, min.dist = 0.1, spread = 0.5)
#min.dist let's me play around with the thightnes of the cells within the clusters
#the spread let's me play around with the coordinates of the distance between the clusters 
##default settings are -> min.dist: 0.3, spread: 1
#se.dc.f <- RunUMAP(se.dc.f, reduction = "pca",dims = 1:10, min.dist = 0.01, spread = 1)

umap.c.cellclass <- DimPlot(se.dc, reduction = "umap", group.by = "cell_class", cols = c.cellclass)
umap.c.clusters <- DimPlot(se.dc, reduction = "umap", group.by = "seurat_clusters")

p.split.all <- DimPlot(se.dc, reduction = "umap", split.by = "cell_class")

cowplot::plot_grid(umap.c.cellclass, umap.c.clusters)
#Export plots to inspect

png(filename = "~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/umaps-minus-lps-lys.png", width = 35, height = 15, res = 300, units = "cm") #Units are pixels by default
print(cowplot::plot_grid(umap.c.cellclass, umap.c.clusters,ncol = 2))
dev.off()

#Split-graphs


png(filename = "~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/umaps-splitbycond-combined-plus-minus-lps-lys.png", width = 60, height = 20, res = 300, units = "cm") #Units are pixels by default
print(cowplot::plot_grid(p.split.all,p.split.filter))
dev.off()

umap.c.cellclass
umap.c.clusters


```
*Cluster 3 is most specific to control, while cluster 2 seems to be mostly specific to be specific to the 3 hour time-point and cluster 3 to the 12 h time-point: look into pairwise DEG between strains

cluster 7 mainly contains LDM at 3h 

cluster 1 and 4 contain all conditions: pairwise DE for timepoint and/ or strain? 

cluster 5 and 6 are present in all conditions (including the control)


```{r, plot umaps with and without lps and/or lysates, fig.width= 12, fig.height=6}

DimPlot(se.dc, reduction = "umap", group.by = "cell_class")
umap.coordinates.combined.filtered <- as.data.frame(Embeddings(se.dc, reduction = "umap"))

```



```{r, save seuratobjects}

saveRDS(se.dc, "~/Dropbox/shared_ankarklev/Franzi-analysis/seurat-objects/seurat-object-dual-DC-TG-200529")
saveRDS(se.dc.f, "~/Dropbox/shared_ankarklev/Franzi-analysis/seurat-objects/seurat-object-dual-nolps-nolys-DC-TG-200529")

```


```{r, add module scores to visualise on the data, fig.width = 12, fig.height=6}

matdc <- c("Ccr7", "Ccl22", "Serpinb9", "Cd83")
immatdc <- c("Tnf", "Il1a", "Cxcl10", "C5ar1")

se.dc <- AddModuleScore(se.dc, features = immatdc, assay = "SCT", name = "immatureDC")
se.dc <- AddModuleScore(se.dc, features = matdc, assay = "SCT", name = "matureDC")

head( se.dc[[]])

FeaturePlot(se.dc, features = c("matureDC1", "immatureDC1"), blend = T)

cowplot::plot_grid(FeaturePlot(se.dc, features = "matureDC1"), FeaturePlot(se.dc, features = "immatureDC1"))

```
*potentially of interest: what is different between the infected cells in the mature vs immature DCs, i.e. how does infection correlate with/affects maturity and the other way around?*

To answer this question: create a new category based on maturity in the meta-data and potentially run pairwise DE tests for timepoint or strain based on maturity level of the DC?

```{r DGEA}

se.dc <- SetIdent(se.dc, value = "seurat_clusters")

se.dc.markers <- FindAllMarkers(se.dc, min.pct = 0.25, logfc.threshold = 2)

c0.m <- subset(se.dc.markers, cluster == 0)
c1.m <- subset(se.dc.markers, cluster == 1)
c2.m <- subset(se.dc.markers, cluster == 2)
c3.m <- subset(se.dc.markers, cluster == 3)
c4.m <- subset(se.dc.markers, cluster == 4)
c5.m <- subset(se.dc.markers, cluster == 5)
c6.m <- subset(se.dc.markers, cluster == 6)
c7.m <- subset(se.dc.markers, cluster == 7)


c0.m <- subset(c0.m, avg_logFC > 2 & p_val_adj < 0.01) 
#somehow subsetting according to the p-value doesn't work ... 

#Find markers for LPS against control

se.dc <- SetIdent(se.dc, value = "cell_class")

lps.m <- FindMarkers(se.dc, ident.1 = "LPS", min.pct = 0.25, logfc.threshold = 2, only.pos = T)
lps.c.m <- FindMarkers(se.dc, ident.1 = "LPS", ident.2 = "ctrl", min.pct = 0.25, logfc.threshold = 2, only.pos = T)

#Filter both datasets for significance: 

lps.m <- subset(lps.m, p_val_adj < 0.005)
lps.m$gene <- rownames(lps.m)
lps.c.m <- subset(lps.c.m, p_val_adj < 0.005)
lps.c.m$gene <-rownames(lps.c.m)

write.xlsx(lps.m, "~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/upregulated-LPS-all-conditions.xlsx")
write.xlsx(lps.c.m, "~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/upregulated-LPS-control.xlsx")

#external_gene_name <- sapply(strsplit(rownames(se.dc.markers), "-" ), "[[", 1)
#ensembl_gene_id <- sapply(strsplit(rownames(se.dc.markers), "-" ), "[[", 2)
#markerannotation <- data.frame(rownames(se.dc.markers), external_gene_name, ensembl_gene_id) %>% rename(  gene = rownames.se.dc.markers.)


```



```{r harmony, fig.width=8, fig.height=4}

library(harmony)
DefaultAssay(se.dc) <- "SCT"
se.dc <- RunHarmony(se.dc, reduction = "pca", group.by.vars = "cell_class", assay.use = "SCT")

se.dc <- RunUMAP(se.dc, reduction = "harmony", dims = 1:20)
se.dc <- FindNeighbors(se.dc, reduction = "harmony", dims = 1:20)
se.dc <- FindClusters(se.dc)

se.dc <- SetIdent(se.dc, value = "cell_class")
p1 <- DimPlot(se.dc, reduction = "umap")
se.dc <- SetIdent(se.dc, value = "SCT_snn_res.0.8")
p2 <- DimPlot(se.dc, reduction = "umap")
cowplot::plot_grid(p1, p2)

DimPlot(se.dc, reduction = "umap", split.by = "cell_class")

```


```{r , pairwise DE tests}

#Paste the cluster to the respective cell_class of each gene and add as a new column in the metadata 
se.dc$cluster_and_cell_class <- paste0(se.dc$SCT_snn_res.0.8, "_", se.dc$cell_class)
id1 <- "2_LDM_3h"
id2 <- "2_PTG_3h"
#Set idenity
se.dc <- SetIdent(se.dc, value = "cluster_and_cell_class")
pw.markers2 <- FindMarkers(object = se.dc, ident.1 = id1, ident.2 = id2, logfc.threshold = 2)
pw.markers2 <- subset(markers1, p_val_adj < 0.05)
#Set filters! 

pw.markers2

```


Try to run some correlation plots of toxo and mouse genes for specific clusters in the data!
Run with gene-name annotated data-set! 

```{r, corrleation tests}

#Select clusters and plot them against each other either 
#Let's start with clusters 2, 3 and 4 
#Use differential expressed genes, potentially also try with pairwise DEGs!

#run correlation within clusters 
deg.cor <- cor(t(as.matrix(se.dc@assays$SCT@data[c(c1.m$gene), ])))

#Try to correlate the 10 or 20 most deferentially/variable expressed genes in each species and correlate them
#Use the data from the DEG analysis for the individual data-sets

#Create character-vector with all genes of interest (top20 most variable toxo and DC Genes)

top20dc.toxo <- c(top20.all$gene, top20.all.t$ensembl_gene_id)

deg.cor <- cor(t(as.matrix(se.dc@assays$SCT@data[c(c4.m$gene), ])))

corrplot(deg.cor, method = "shade", tl.col = "black", order = "FPC")

#regulatory network
library(qgraph)

png("~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/GeneRegultoryNetwork-prelim.png", width = 30, height = 20, units = "cm", res = 300)
qgraph(deg.cor, layout = "spring", sampleSize = nrow(se.dc@assays$SCT@data))
dev.off()

```


```{r, sources of differences between clusters}

#Number of toxoplasma transcripts/cluster
nc <- c(0:7)
n <- c("sum.c0", "sum.c1", "sum.c2", "sum.c3", "sum.c4", "sum.c5", "sum.c6", "sum.c7")

cluster.t.p.l <- setNames(sapply(nc, function(x){
subset(se.dc, subset = seurat_clusters %in% x)
}), nm = n)

cluster.t.p.l <- setNames(lapply(cluster.t.p.l, function(x){
  sum(x@meta.data$toxo_nUMI_p)
}), nm = n) 

cluster.t.p.l <- as.data.frame(unlist(cluster.t.p.l))

names(cluster.t.p.l)[1] <- "percentage"
cluster.t.p.l$cluster <- as.character(0:7)

#Plot the percentages in a bar-graph 
png("~/Dropbox/shared_ankarklev/Franzi-analysis/Results-FiguresAndTables/relativeAbdundnaceToxotranscriptspercluster.png", width = 30, height = 20, units = "cm", res = 300)
print(ggplot(cluster.t.p.l, aes(cluster, percentage)) + 
  geom_bar(stat = "identity") + 
  theme_classic() + 
  ylab("nUMI T. gondii of total nUMI [%]"))
dev.off()


dc.only.l <- setNames(lapply(dc.only.markers.l, function(x){
  x$ensembl_gene_id <- rownames(x)
  x <- dplyr::inner_join(x, genes.table, by = "ensembl_gene_id")
}) , nm = dc.only.n)



```

```{r, correlate the number of transcripts with the abundance of specific genes (possibly host genes) in the cluster}

#Set identity differently to look at these correlations instead of cluster-wise! 

FeatureScatter(se.dc, feature1 = "nFeature_RNA", feature2 = "nCount_RNA")


```

```{r, AddModuleScore based on maturity}


matdc <- c("Ccr7", "Ccl22", "Serpinb9", "Cd83")
immatdc <- c("Tnf", "Il1a", "Cxcl10", "C5ar1")

se.dc <- AddModuleScore(se.dc, features = matdc, assay = "SCT", name = "matureDC", seed = 1)

head( se.dc[[]])

#ModuleScores added for each gene of each cell -> 
#write if statement that says if 3 our of 4 are positive = mature, if 2 out of 4 = positive = intermediate, if 1 out of 4 = positive = immature! 

maturedcs1 <- sapply((se.dc@meta.data$matureDC1), function(x) {
  if (x > 0)  paste0(1) else paste0(0)
  })
maturedcs2 <- sapply((se.dc@meta.data$matureDC2), function(x) {
  if (x > 0)  paste0(1) else paste0(0)
  })
maturedcs3 <- sapply((se.dc@meta.data$matureDC3), function(x) {
  if (x > 0)  paste0(1) else paste0(0) 
  })

maturedcs4 <- sapply((se.dc@meta.data$matureDC4), function(x) {
  if (x > 0)  paste0(1) else paste0(0) 
  })

maturityscores <- data.frame(maturedcs1, maturedcs2, maturedcs3, maturedcs4, stringsAsFactors = F) %>% 
  mutate_all(as.numeric) 
rownames(maturityscores) <- rownames(se.dc@meta.data)
maturityscores <- transform(maturityscores, sum = rowSums(maturityscores))

#attach maturity as a column to the dataframe: 

maturityscores<- as.data.frame(sapply((maturityscores$sum), function(x) {
  if(x >= 3) paste0("mature") 
  else paste0("immature")
  }))

rownames(maturityscores) <- rownames(se.dc@meta.data)

se.dc <- AddMetaData(se.dc, metadata = maturityscores, col.name = "maturation_state")

DimPlot(se.dc, group.by = "cell_class")

FeaturePlot(se.dc, features = "Ccr7")

```

```{r, pairwise DGE between mature, immature}


se.dc <- SetIdent(se.dc , value = "cell_class_maturation")

maturation.markers <- FindAllMarkers(se.dc, min.pct = 0.25, logfc.threshold = 0.5, only.pos = T)
maturation.markers <- subset(maturation.markers, p_val_adj < 0.05)
#Run pairwise DGE based on maturation and cluster/cell_class

#Paste the cluster to the respective cell_class of each gene and add as a new column in the metadata 
se.dc$cluster_maturation <- paste0(se.dc$SCT_snn_res.0.8, "_", se.dc$maturation_state)
se.dc$cell_class_maturation <- paste0(se.dc$cell_class, "_", se.dc$maturation_state)

#Create IDs fro groups of interest we would like to observe in more detail 

id1 <- "LDM_3h_immature"
id2 <- "ctrl_immature"
#Set idenity
#se.dc <- SetIdent(se.dc, value = "cluster_and_cell_class")
pw.markers2 <- FindMarkers(object = se.dc.mus, ident.1 = id1, ident.2 = id2, logfc.threshold = 2)
pw.markers2 <- subset(pw.markers2, p_val_adj < 0.05)
#Set filters! 

#Make heatmap with DEGs based on this analysis 


```

```{r, add toxoclusters to metadata and highlight them in umap}

toxo_clusters <- as.data.frame(paste0("toxo", "_", se.toxo$SCT_snn_res.0.8))
rownames(toxo_clusters) <- rownames(se.toxo[[]])
names(toxo_clusters)[1] <- "toxo_cluster"
head(toxo_clusters)

dc_clusters <- as.data.frame(paste0("dc", "_", se.dc.mus$SCT_snn_res.0.8))
rownames(dc_clusters) <- rownames(se.dc.mus[[]])
names(dc_clusters)[1] <- "dc_cluster"
head(dc_clusters)

se.dc <- AddMetaData(se.dc, metadata = toxo_clusters, col.name = "toxoclusters")
se.dc <- AddMetaData(se.dc, metadata = dc_clusters, col.name = "dcclusters")
#Visulize only cells of toxo_cluster = toxo_2 on DC uMAP!

se.dc <- SetIdent(se.dc, value = "toxoclusters")

toxo.dc.0 <- DimPlot(se.dc, cols = c("darkred", "gray48", "gray48", "gray48", "gray48", "gray48", "gray48"))

toxo.dc.1 <- DimPlot(se.dc, cols = c("gray48", "darkred", "gray48", "gray48", "gray48", "gray48", "gray48"))

toxo.dc.2 <- DimPlot(se.dc, cols = c("gray48", "gray48","darkred" ,"gray48", "gray48", "gray48", "gray48"))
  
toxo.dc.3 <- DimPlot(se.dc, cols = c("gray48",  "gray48", "gray48", "darkred","gray48", "gray48", "gray48"))

toxo.dc.4 <- DimPlot(se.dc, cols = c("gray48",  "gray48", "gray48","gray48",  "darkred","gray48", "gray48"))

cowplot::plot_grid(toxo.dc.0, toxo.dc.1, toxo.dc.2, toxo.dc.3, toxo.dc.4)

se.dc <- SetIdent(se.dc, value = "dcclusters")

dc.0 <- DimPlot(se.dc, cols = c("darkblue", "gray48", "gray48", "gray48", "gray48", "gray48", "gray48","gray48", "gray48" ))

dc.1 <- DimPlot(se.dc, cols = c("gray48", "darkblue", "gray48", "gray48", "gray48", "gray48", "gray48","gray48", "gray48"))

dc.2 <- DimPlot(se.dc, cols = c("gray48", "gray48","darkblue" ,"gray48", "gray48", "gray48", "gray48","gray48", "gray48"))
  
dc.3 <- DimPlot(se.dc, cols = c("gray48",  "gray48", "gray48", "darkblue","gray48", "gray48", "gray48","gray48", "gray48"))

dc.4 <- DimPlot(se.dc, cols = c("gray48",  "gray48", "gray48","gray48",  "darkblue","gray48", "gray48","gray48", "gray48"))

dc.5 <- DimPlot(se.dc, cols = c("gray48", "gray48", "gray48", "gray48", "gray48", "darkblue", "gray48","gray48", "gray48"))

dc.6 <- DimPlot(se.dc, cols = c("gray48", "gray48", "gray48", "gray48", "gray48", "gray48", "darkblue","gray48", "gray48"))

DimPlot(se.dc, group.by = "cell_class")
FeaturePlot(se.dc, features = c("Il1b", "Il1a"), blend = T)

png("~/Dropbox/shared_ankarklev/Franzi-analysis/preliminary-umap-combined-dc-clusters-seperated.png", width = 30, height = 20, unit= "cm", res = 300)
print(cowplot::plot_grid(dc.0, dc.1, dc.2, dc.3, dc.4, dc.5, dc.6))
dev.off()

png("~/Dropbox/shared_ankarklev/Franzi-analysis/preliminary-umap-combined-toxo-clusters-seperated.png", width = 30, height = 20, unit= "cm", res = 300)
cowplot::plot_grid(toxo.dc.0, toxo.dc.1, toxo.dc.2, toxo.dc.3, toxo.dc.4)
dev.off()

png("~/Dropbox/shared_ankarklev/Franzi-analysis/preliminary-umap-combined-dc-clusters.png", width = 30, height = 20, unit= "cm", res = 300)
DimPlot(se.dc, group.by = "dcclusters")
dev.off()

png("~/Dropbox/shared_ankarklev/Franzi-analysis/preliminary-umap-combined-toxo-clusters.png", width = 30, height = 20, unit= "cm", res = 300)
DimPlot(se.dc, group.by = "toxoclusters")
dev.off()

png("~/Dropbox/shared_ankarklev/Franzi-analysis/preliminary-umap-combined-cellclass.png", width = 30, height = 20, unit= "cm", res = 300)
DimPlot(se.dc, group.by = "cell_class")
dev.off()

```