---
title: "Toxo_screen_analysis"
output: html_document
date: '2022-04-04'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Saeij et al. published a genome wide CRISPR screen to find fitness related genes in Toxoplasma infected Macrophages. We would like to compare their results to our data and see if there is mouse cluster specific expression of the fitter Toxoplasma infected cells and also to see if we find correlated upregulation of genes or genes of interest in our network.

First we load the data for the toxoplasma crispr hits
```{r , load data}

library(readxl)
crisp <- as.data.frame(read_excel("~/t.gondi_dc/data/CRISPR_screen_data_doi-org.ezp.sub.su.se:10.1038:s41467-020-18991-8/41467_2020_18991_MOESM5_ESM.xlsx", 
    sheet = "Fitness"))
head(crisp)

crisp_naive_vs_hff <- subset(crisp,crisp[,8] == "YES")
crisp_naive_vs_hff <- crisp_naive_vs_hff[,c(2:11)]
crisp_naive_vs_hff$ME49_ID <- gsub("_", "-",crisp_naive_vs_hff$ME49_ID)
head(crisp_naive_vs_hff)

crisp_INF_vs_naive <-  subset(crisp,crisp[,12] == "YES")
crisp_INF_vs_naive <- crisp_INF_vs_naive[,c(2:13)]
#crisp_INF_vs_naive$ME49_ID <- gsub("_", "-",crisp_INF_vs_naive$ME49_ID)
head(crisp_INF_vs_naive)

nrow(crisp_naive_vs_hff)
nrow(crisp_INF_vs_naive)

FeaturePlot(se.toxo, features = crisp_naive_vs_hff$ME49_ID)
FeaturePlot(se.toxo, features = crisp_INF_vs_naive$ME49_ID)

```

## Expanded correlation network analysis 

```{r, calcualte correlations}

#write function to calculate correlation of gene sets of interest 
set.seed(3)
#mat1 = genenames as rows and cells as cols
#mat2 = genenames as rows and cells as cols 
cor.tx <- function(mat1,mat2, thrs.pos = 0.2, thrs.neg = -0.2) {
  

cor1 = bicor(t(mat1),t(mat2),use = "pair")


cor1[1:5,1:5]
top.cor <- data.frame(mean_cor = rowMeans(cor1,na.rm = T))
top.cor$gene <- rownames(top.cor)
top.cor <- top.cor[order(top.cor$mean_cor,decreasing = T), ]

p <- ggplot(top.cor,aes(reorder(gene, mean_cor),mean_cor)) + ## plot to decide for a meaningful cut-off 
  geom_jitter()
#based on the results we first set a cut-off of 0.1 for positive correlations and -0.1 for negative correlations 
top.cor_pos <- top.cor[top.cor$mean_cor > thrs.pos,]
nrow(top.cor_pos)
##we can check which genes are negatively correlated with rop-expression 
top.cor_neg <- top.cor[top.cor$mean_cor < thrs.neg,]
nrow(top.cor_neg)
return(list(p, as.data.frame(rbind(top.cor_pos, top.cor_neg))))

}


```

```{r, make correlation matrix with the extracted genes}

genes.table.tg$mat_id <- gsub("_", "-", rownames(genes.table.tg))
set.seed(3)

corplot.tx <- function(data, seed, annotation,method, order, clust = clust){
#make correlation matrix for selected genes 
cor.mat.1 <- as.matrix(data[seed,])
#attach genes names (if present)
  annotation <- subset(annotation, annotation$ensembl_gene_id %in% intersect(rownames(cor.mat.1), annotation$ensembl_gene_id))
  cor.mat.1 <- cor.mat.1[annotation$ensembl_gene_id,]
  rownames(cor.mat.1) <- paste0(annotation$ensembl_gene_id,"::",annotation$external_gene_name)
  rownames(cor.mat.1) <- gsub("::N/A", "", rownames(cor.mat.1))
  cor.mat.1 <- t(cor.mat.1)
  cor.mat.1 <- cor(cor.mat.1, method = "pearson")

return(list(cor.mat.1,corrplot(cor.mat.1, method = method,type = "upper",order = order, clust.method = clust, tl.col = "black",tl.cex = 0.6, col = colorRampPalette(c("#660156","#21908C", "white", "#FFF68F", "#FFD700"))(1001))))
}

corm.tx <- function(data, seed, annotation){
#make correlation matrix for selected genes 
set1 <- union(seed, rownames(data))
cor.mat.1 <- t(as.matrix(data[set1,]))
#attach genes names (if present)
tab <- subset(annotation, annotation$mat_id %in% seed)
cor.mat.1 <- cor(cor.mat.1, method = "pearson")
tab %>%
  arrange(factor(mat_id, levels = rownames(cor.mat.1)))
rownames(cor.mat.1) <- tab$combined
rownames(cor.mat.1) <- gsub("N/A", "hypothetical protein", rownames(cor.mat.1))
return(cor.mat.1)

}

```

Now we need to check which genes of our correlation networks are part of the S/M, G1 sub-transcriptomes, see if we observe enrichment in one of the cell cycle phases in our data and potentially remove them from the downstream analysis. 

```{r, analyze cell cycle/replication influence}

#look at overlap of tab1 (interaction cluster 1 overlap)

cell_cycle_tab9 <- c(intersect(tab9.tg$ensembl_gene_id, G1.dat$ensembl_gene_id),intersect(tab9.tg$ensembl_gene_id, S_M.dat$ensembl_gene_id))

cell_cycle_tab1 <- c(intersect(tab1.tg$ensembl_gene_id, G1.dat$ensembl_gene_id),intersect(tab1.tg$ensembl_gene_id, S_M.dat$ensembl_gene_id))

#Now we remove these genes from the tab9 and tab1 lists

tab9.tg <- tab9.tg[-(which(tab9.tg$ensembl_gene_id %in% cell_cycle_tab9)),]
tab1.tg <- tab1.tg[-(which(tab1.tg$ensembl_gene_id %in% cell_cycle_tab1)),]

```


```{r, investigate the correlation between interaction cluster genes and other Toxogenes}

##Subset remaining expression matrix to not contain any cell cycle/replication marker genes and explore correlation 
M_ncc <- M_toxo[-which(c(rownames(M_toxo) %in% G1.dat$ensembl_gene_id, rownames(M_toxo) %in% S_M.dat$ensembl_gene_id)),]

#calculate the correlations (bicor) between the virulent and the maintance cluster and exclude all genes with average positive  correlations (Rowsums) below 0.25 or negative correlations above -0.25
tab1.tg_genes <- tab1.tg$ensembl_gene_id
int.1_m <- M_ncc[rownames(M_ncc) %in% tab1.tg_genes,]
inter.1.cor <- cor.tx(mat1 = M_ncc, mat2 = int.1_m, thrs.pos = 0.17, thrs.neg = -0.17)
tab9.tg_genes <- tab9.tg$ensembl_gene_id
int.9_m <- M_toxo[rownames(M_toxo) %in% tab9.tg_genes,]
inter.9.cor <- cor.tx(mat1 = M_toxo, mat2 = int.9_m, thrs.pos = 0.25, thrs.neg = -0.25)

#Run correlation analysis 

#pdf("~/t.gondi_dc/res/figures/fig4/inter1_correlation_pearson_ncc.pdf")
corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab1.tg$ensembl_gene_id), rownames(inter.1.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[2]]
#dev.off()

write.table(corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab1.tg$ensembl_gene_id), rownames(inter.1.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[1]], sep = ";", row.names = rownames(corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab1.tg$ensembl_gene_id), rownames(inter.1.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[1]]), file = "~/t.gondi_dc/res/supplementary_data/supplementary_data_5.csv", quote = F)

#pdf("~/t.gondi_dc/res/figures/fig4/inter9_correlation_pearson_ncc.pdf")
corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab9.tg$ensembl_gene_id), rownames(inter.9.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[1]]
#dev.off()

write.table(corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab9.tg$ensembl_gene_id), rownames(inter.9.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[1]], sep = ";", row.names = rownames(corplot.tx(data = se.toxo@assays$SCT@counts, seed = c(unique(tab9.tg$ensembl_gene_id), rownames(inter.9.cor[[2]])), annotation = genes.table.tg, method = "square",order = "AOE", clust = "average")[[1]]),file ="~/t.gondi_dc/res/supplementary_data/supplementary_data_6.csv", quote = F )

```

```{r, extract only correlated genes and look at the gene descriptions and possibly Gene Ontologies}


all_genes <- c(gsub("-","_",rownames(inter.1.cor)),gsub("-","_",unique(tab1.tg$ensembl_gene_id)),gsub("-","_",rownames(inter.9.cor)),gsub("-","_",unique(tab9.tg$ensembl_gene_id)))
write.table(all_genes,"~/t.gondi_dc/res/gene-list-interaction_tg.txt", quote = FALSE, sep = ",", row.names = F, col.names = F)


go.table.int <- read.table("~/t.gondi_dc/data/meta/go_list_toxo_interaction.csv", sep =",", header = T)
go.table.int[go.table.int == "N/A"] <- NA

#Make aggregated table 

go.table.int <- go.table.int %>% 
  gather(key = "Category", value = "GO_ID", c(5,7,10,12,14,16)) %>%
  gather(key = "Category", value = "GO_Description", c(5:10)) 

comp <- grep("Computed",go.table.int$Category)
go.table.int$curation <- NA
go.table.int[comp,]$curation <- paste0("Computed")
go.table.int[-comp,]$curation <- paste0("Curated")
go.table.int[c('Parent', 'First', 'Last')] <- str_split_fixed(go.table.int$GO_Description, ';', 3)
go.table.int$ID_Description <- paste0(go.table.int$GO_Description, "::", go.table.int$GO_ID)
go.table.int[go.table.int == ""] <- NA


go.table.int[go.table.int$Gene.ID %in% gsub("-","_",rownames(inter.1.cor[[2]])),]
go.table.int[go.table.int$Gene.ID%in% gsub("-","_",unique(tab1.tg$ensembl_gene_id)),]

go.table.int[go.table.int$Gene.ID%in% gsub("-","_",rownames(inter.9.cor[[2]])),]
go.table.int[go.table.int$Gene.ID %in% gsub("-","_",unique(tab9.tg$ensembl_gene_id)),]


#cat =  one of : "Computed.GO.Components" "Computed.GO.Functions"  "Computed.GO.Processes"  "Curated.GO.Components"  "Curated.GO.Functions"   "Curated.GO.Processes" 
#col = one of : "GO_Description" , "GO_ID" "ID_Description"

#convert category and column into factors
go.table.int$Category <- as.factor(go.table.int$Category)
go.table.int$GO_ID <- as.factor(go.table.int$GO_ID)
go.table.int$GO_Description <- as.factor(go.table.int$GO_Description)
go.table.int$ID_Description <- as.factor(go.table.int$ID_Description)


#fraction(divided by number of genes) of each term in geneset/fraction of each term in whole data 
#statstics: fisher + bonferoni /bootstrap

pie_toxogo <- function(df, seed,nterm = c(1:nrow(count.table)),cols, cat, col) {

df$Category <- as.factor(df$Category)
df <- subset(df, df$Category == cat)

df.1 <- df[df$Gene.ID %in% seed,]
count.table <- count(df.1, df.1[,col])
count.table <- na.omit(count.table)
colnames(count.table) <- c("Term", "count")

count.table <- count.table[order(-count.table$count),]
count.table$Term <- as.factor(count.table$Term)

p <- ggplot(count.table[nterm,], aes(y= reorder(Term, count),x=count, fill = Term)) +
    geom_bar(#width = 1,
             stat = "identity"
             ) +
    #scale_y_continuous(breaks = 0:nlevels(count.table$Term)) +
    #scale_color_gradientn(colors = colorRampPalette(c("#fdfdfc" ,"#eeeeee", "#adafaa", "#666d70", "black"))(1000)) +
    scale_fill_manual(values = colorRampPalette(cols)(100)) +
                      
                      #c("#f9c80e", "#f86624", "#ea3546", "#662e9b", "#43bccd", "#0466c8", "#bfd200", "#548c2f", "#41ead4", "#6a040f", "#8ac926", "#A44200", "#023047", "#b5179e", "#ef476f", "#6b9080", "#e09f3e", "#e0aaff", "#03045e", "#e0b1cb", "#aaf683")) + 
    theme_bw() +
    theme(axis.ticks = element_blank(),
          #axis.text = element_blank(),
          axis.title = element_blank(),
          axis.line = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(), 
          legend.position = "none", 
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank()
          ) 
          #coord_polar() 
    
     

return(list(count.table,p))

}


seed1_full = gsub("-","_",c(unique(tab1.tg$ensembl_gene_id), rownames(inter.1.cor[[2]])))
seed2_full = gsub("-","_",c(unique(tab9.tg$ensembl_gene_id), rownames(inter.9.cor[[2]])))
seed1 = gsub("-","_",c(rownames(inter.1.cor[[2]])))
seed2 = gsub("-","_",c(rownames(inter.9.cor[[2]])))
#cat =  one of : "Computed.GO.Components" "Computed.GO.Functions"  "Computed.GO.Processes"  "Curated.GO.Components"  "Curated.GO.Functions"   "Curated.GO.Processes" 

#col = one of : "GO_Description" , "GO_ID" "ID_Description", "Parent", "First", "Last"

pdf("~/t.gondi_dc/res/figures/fig4/enrich_1_function.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, nterm = c(1:5),seed1_full, cols = c("#d64701", "#ffd7c6"), cat = "Computed.GO.Functions", col = "Parent")
dev.off()

pdf("~/t.gondi_dc/res/figures/fig4/enrich_1_process.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, seed1_full, cols = c("#d64701", "#ffd7c6"),cat = "Computed.GO.Processes", nterm = c(1:5), col = "Parent")
dev.off()

pdf("~/t.gondi_dc/res/figures/fig4/enrich_1_component.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, seed1_full, cols = c("#d64701", "#ffd7c6"),cat = "Computed.GO.Components",nterm = c(1:5), col = "Parent")
dev.off()

pdf("~/t.gondi_dc/res/figures/fig4/enrich_9_function.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, seed2_full,cols = c("#140e6a","#ffd7c6"),cat = "Computed.GO.Functions", col = "Parent")
dev.off()

pdf("~/t.gondi_dc/res/figures/fig4/enrich_9_process.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, seed2_full,cols = c("#140e6a","#ffd7c6"),cat = "Computed.GO.Processes", col = "Parent")
dev.off()

pdf("~/t.gondi_dc/res/figures/fig4/enrich_9_component.pdf", width = 10, height = 2)
pie_toxogo(go.table.int, seed2_full, cols = c("#140e6a","#ffd7c6"),cat = "Computed.GO.Components", col = "Parent")
dev.off()


```


##CRISPR comparison 

```{r, averaged gene expression across cells of toxo genes (grouped by clusters), fig.width = 5, fig.height=10}


data <- se.toxo@assays$SCT@counts
dat <- t(as.matrix(data))

hmat_t <- hmat_mousetoxo(data = dat, meta = se.toxo$seurat_clusters)
hmat_c <- hmat_mousetoxo(data = dat, meta = se.toxo$cell_class)
hmat_c <- as.data.frame(hmat_c)
hmat_c <- hmat_c %>% select("PTG_3h", "LDM_3h", "PTG_12h", "LDM_12h")
hmat_c <- as.matrix(hmat_c)
hmat_m <- hmat_mousetoxo(data = dat, meta = se.toxo$mouseclusters)

# function to get heatmaps 
t_hmat_crisp <- function(hmat, seed) {
  
  seed$ensembl_gene_id <- gsub("_", "-",seed$ensembl_gene_id)
  seed <- subset(seed, seed$ensembl_gene_id %in% intersect(rownames(hmat), seed$ensembl_gene_id))
  hmat_rop <- hmat[seed$ensembl_gene_id,]
  rownames(hmat_rop) <- paste(seed$ensembl_gene_id, "::",seed$`Gene products description`)
  p <- pheatmap(hmat_rop, fontsize_row = 8, cellwidth = 15, clustering_distance_rows = "correlation", breaks = seq(-max(abs(hmat_rop)), max(abs(hmat_rop)), length.out = 1000), color = colorRampPalette(c("#660156","#21908C", "white", "#FFF68F", "#FFD700"))(1001), angle_col = "45", cluster_rows = T, cluster_cols = F) 
  
  return(p)
}

#get the clustering
crisp_INF_vs_naive$ensembl_gene_id <- gsub("_", "-",crisp_INF_vs_naive$ME49_ID)


pdf("~/t.gondi_dc/res/figures/fig4/Crispr_INF_vs_naive_hmat_conditions_new.pdf")
t_hmat_crisp(hmat_c,crisp_INF_vs_naive)
dev.off()

```


```{r, visualize gene sets in co-expression network}


#interaction networks for T.gondii fitness after INF stimulation

##MAKE CHANGES to the toxo networks to color genes that are in the original data and make remaining grey but keep mouse genes all in one color (should be opposite of what is in the current network for mouse)
gr.inf <- f.KNN_toxo(crisp_INF_vs_naive$ensembl_gene_id, Nsteps = 3, K = 3, cutoffValue = 0.3
                          )

pdf("~/t.gondi_dc/res/figures/supplementary/t.gondii_INF_network.pdf", width = 12, height = 12)
gr.inf <- f.KNN_toxo(crisp_INF_vs_naive$ensembl_gene_id, Nsteps = 3, K = 3, cutoffValue = 0.3
                          )
dev.off()

#Get the data per co-expression cluster
tab <- get.toxo.table(gr.inf)
##Find out largest clustrs 
tab$`co-expression_cluster` <- as.factor(tab$`co-expression_cluster`)
table(tab$`co-expression_cluster`)

#print cluster 3 + 4 as they are the two biggest clusters 
pdf("~/t.gondi_dc/res/figures/t.gondii_INF_network_sub3.pdf", width = 12, height = 5)
subgraph_mtoxo_clust(gr.inf, clust = 3)
dev.off()
pdf("~/t.gondi_dc/res/figures/t.gondii_INF_network_sub4.pdf", width = 12, height = 5)
subgraph_mtoxo_clust(gr.inf, clust = 4)
dev.off()

##check larger networks with more nodes and co-expressed genes (not in manuscript)
gr.inf.large <- f.KNN_toxo(crisp_INF_vs_naive$ensembl_gene_id, Nsteps = 50, K = 2 , cutoffValue = 0.2)




```


