---
title: "Analysis not used in manuscript"
output: html_notebook
---

In addition: Try to run the DGEA according to the original cell-identity and look at the results and the GO-term analysis in the DCs! 


```{r, Clusterprofiler analysis for cluster 4}


markers_4 <- FindMarkers(se.dc.mus, ident.1 = "M 7") #refers to LDM infection at 12h
markers_4 <- subset(markers_4, p_val_adj < 0.05)
original.gene.list <- markers_4$avg_log2FC
markers_4$gene <- rownames(markers_4)
names(original.gene.list) <- markers_4$gene
gene_list<-na.omit(original.gene.list)
gene_list = sort(gene_list, decreasing = TRUE)
head(gene_list)

gseM7 <- gseGO(geneList=gene_list, 
             ont ="BP",
             keyType = "SYMBOL",
             minGSSize = 3, 
             maxGSSize = 500, 
             pvalueCutoff = 0.1, 
             pAdjustMethod = "hochberg",
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db)

head(gseM7@result)


#select the most up/downregulated pathways/pathways of interest

ridgeplot(gse4, showCategory = 20,  core_enrichment = T) + 
  ggplot2::labs(x = "enrichment distribution") +
  ggplot2::scale_fill_gradientn(colors = colorRampPalette(c("#440154" ,"#21908C", "#FDE725"))(200))

#png("~/t.gondi_dc/res/figures/fig3/gse_genes_infected12h_enrichment.png", width = 30, height = 25, res = 600, units = "cm")
dotplot(gse4, showCategory= 10, split=".sign") + 
  ggplot2::facet_grid(.~.sign) +
  ggplot2::scale_color_gradientn(colors = colorRampPalette(c("#FDE725", "#21908C","#440154" ))(200))
#dev.off()


cnetplot(gse4, categorySize="pvalue",foldChange = gene_list)

p <- heatplot(gseM7, showCategory = 15, foldChange = gene_list) +
  ggplot2::scale_fill_gradientn(colors = colorRampPalette(c("#440154" ,"#21908C", "#FDE725"))(200))

edox2 <- pairwise_termsim(gseM6)
treeplot(edox2) + 
  scale_color_gradientn("p-value",colors = colorRampPalette(c("#660156" ,"#21908C", "#FDE725"))(200))

edox <- pairwise_termsim(gseM7)
treeplot(edox) + 
  scale_color_gradientn("p-value",colors = colorRampPalette(c("#660156" ,"white", "#FDE725"))(200))


#png("~/t.gondi_dc/res/figures/fig3/gse_genes_12h.png", width = 25, height = 12, units = "cm", res = 600)
print(p)
#dev.off()


```

If of interest we can have a look into the late infected GM-DCs in particular (not shown in the manuscript)

```{r, Clusterprofiler analysis for cluster M2 (late infected dendritic cells)}

se.dc.mus <- SetIdent(se.dc.mus, value = "seurat_clusters")
markers_m2 <- FindMarkers(se.dc.mus, ident.1 = "M 2", logfc.threshold = 0.25)
markers_m2 <- subset(markers_m2, p_val_adj < 0.05)
original.gene.list <- markers_m2$avg_log2FC
markers_m2$gene <- rownames(markers_m2)
names(original.gene.list) <- markers_m2$gene
gene_list<-na.omit(original.gene.list)
gene_list = sort(gene_list, decreasing = TRUE)
head(gene_list)

gseM2 <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "SYMBOL",
             exponent = 1, 
             minGSSize = 3,
             pvalueCutoff = 0.01, 
             verbose = TRUE, 
             OrgDb = org.Mm.eg.db, 
             pAdjustMethod = "hochberg", 
             by = "fgsea")


#supplement add treeplot with all GO-terms 
edox2 <- pairwise_termsim(gseM2)

#png("~/t.gondi_dc/res/figures/fig3/treeplot_M2.png", width = 47, height = 25, units = "cm", res = 600)
treeplot(edox2) + 
  scale_color_gradientn("p-value",colors = colorRampPalette(c("#660156" ,"#21908C", "#FDE725"))(200))
#dev.off()


#Manual selection of terms of interest
gseM2@result <- subset(gseM2@result, gseM2@result$Description %in% c("DNA metabolic process",
                                                                     "regulation of cell motility",
                                                                     "defense response to bacterium", 
                                                                     "positive regulation of cytokine production", 
                                                                     "cell communication"
  
))

```

```{r, run GO term GO term analyis based on clusters}

dc.markers.C0 <- subset(se.dc.markers, se.dc.markers$cluster == "M 1")
dc.markers.C1 <- subset(se.dc.markers, se.dc.markers$cluster == "M 2")
dc.markers.C2 <- subset(se.dc.markers, se.dc.markers$cluster == "M 3")
dc.markers.C3 <- subset(se.dc.markers, se.dc.markers$cluster == "M 4")
dc.markers.C4 <- subset(se.dc.markers, se.dc.markers$cluster == "M 5")
dc.markers.M6 <- subset(se.dc.markers, se.dc.markers$cluster == "M 6")
dc.markers.M7 <- subset(se.dc.markers, se.dc.markers$cluster == "M 7")
dc.markers.C7 <- subset(se.dc.markers, se.dc.markers$cluster == "M 8")
dc.markers.C8 <- subset(se.dc.markers, se.dc.markers$cluster == "M 9")


marker.list <- list(dc.markers.C0, dc.markers.C1, dc.markers.C2, dc.markers.C3, dc.markers.C4, dc.markers.C5, dc.markers.C6, dc.markers.C7, dc.markers.C8)

dc.markers.n <- c("dc.markers.C0","dc.markers.C1", "dc.markers.C2", "dc.markers.C3","dc.markers.C4", "dc.markers.C5", "dc.markers.C6", "dc.markers.C7", "dc.markers.C8")

##get markers that are unique to Cluster M5 

M5_toxo_unique <- FindMarkers(se.dc.mus,only.pos = F, ident.1 = "M 5", ident.2 = c("M 1", "M 2", "M 3", "M 4", "M 6", "M 7"))
M5_toxo_unique <- M5_toxo_unique[M5_toxo_unique$p_val_adj <= 0.01,]
M5_toxo_unique <- M5_toxo_unique[order(-M5_toxo_unique$avg_log2FC),]
M5_toxo_u_pos <- M5_toxo_unique[M5_toxo_unique$avg_log2FC > 0,]
M5_toxo_u_neg <- M5_toxo_unique[M5_toxo_unique$avg_log2FC < 0,]

goM5_pos <- gost(query = rownames(M5_toxo_u_pos), organism = "mmusculus", sources = "WP", significant = T)
goM5_neg <- gost(query = rownames(M5_toxo_u_neg), organism = "mmusculus", sources = "WP", significant = T)

print(ggplot(head(goM5_pos$result, 5), aes(reorder(term_name, -log10(p_value)), -log10(p_value))) +
  geom_bar(stat = "identity", color = "black", fill = "darkblue") +
  coord_flip() +
  #scale_fill_gradientn(name = "Enrichment Score",colours =   RColorBrewer::brewer.pal(n = 20, name = "Reds")) + 
  scale_y_continuous() +
  xlab("Biological Process")+
  ylab("Enrichment Score") +
  theme(axis.text.y = element_text(size = 12), axis.title = element_blank(), panel.background = element_blank()))
goM5 <- gost(query = rownames(M5_toxo_u_pos), organism = "mmusculus", sources = "REAC")

print(ggplot(head(goM5_neg$result, 5), aes(reorder(term_name, -log10(p_value)), -log10(p_value))) +
  geom_bar(stat = "identity", color = "black", fill = "darkred") +
  coord_flip() +
  #scale_fill_gradientn(name = "Enrichment Score",colours =   RColorBrewer::brewer.pal(n = 20, name = "Reds")) + 
  scale_y_continuous() +
  xlab("Biological Process")+
  ylab("Enrichment Score") +
  theme(axis.text.y = element_text(size = 12), axis.title = element_blank(), panel.background = element_blank()))


subset.markers.l <- setNames(lapply(marker.list, function(x) {
  subset(x, avg_log2FC > 0.5 & p_val_adj < 0.01)
}), nm = dc.markers.n)

```

We can also determine whether a cell in the data is positive or negative for the expression of a certain gene of interest, for example Rop16 and Gra15 which are known to trigger different immune pathways in T. gondii infected cells and are dependent on the parasite (not shown in the manuscript). 

```{r, visualize the cells that are Gra15+ and Rop16+ in the DC set}

toxo_gra <- as.data.frame(paste0(se.toxo$Gra15_expression))
rownames(toxo_gra) <- rownames(se.toxo[[]])
names(toxo_gra)[1] <- "gra-expression"
head(toxo_gra)

se.dc.mus <- AddMetaData(se.dc.mus, metadata = toxo_gra, col.name = "Gra15_expression")


toxo_rop <- as.data.frame(paste0(se.toxo$Rop16_expression))
rownames(toxo_rop) <- rownames(se.toxo[[]])
names(toxo_rop)[1] <- "rop-expression"
head(toxo_rop)

se.dc.mus <- AddMetaData(se.dc.mus, metadata = toxo_rop, col.name = "Rop16_expression")

DimPlot(se.dc.mus, group.by = "Rop16_expression")

se.dc.rop <- subset(se.dc.mus, subset = Rop16_expression %in% "Rop16+", slot = "data")
se.dc.gra <- subset(se.dc.mus, subset = Gra15_expression %in% "Gra15+", slot = "data")
DimPlot(se.dc.rop, group.by = "toxoclusters")
DimPlot(se.dc.gra, group.by = "cell_class")

DimPlot(se.dc.mus, group.by = "Gra15_expression")
DimPlot(se.dc.mus, group.by = "Rop16_expression")

```

We want to run RNA velocity experiments using scvelo in python - to this end we are exporting the metadata of our seurat object and format it into an adata object using this tutorial: https://smorabit.github.io/tutorials/8_velocyto/

```{r, write an adata object for velocity experiments}

se.dc.mus$barcode <- colnames(se.dc.mus)
se.dc.mus$UMAP_1 <- se.dc.mus@reductions$umap@cell.embeddings[,1]
se.dc.mus$UMAP_2 <- se.dc.mus@reductions$umap@cell.embeddings[,2]

# save metadata table:
write.csv(se.dc.mus@meta.data, file = "~/t.gondi_dc/data/adata_create/metadata.csv", quote = F, row.names = F)

# save metadata table:library(Matrix)
counts_matrix <- GetAssayData(se.dc.mus, assay = "RNA", slut = "counts")

# write expression counts matrix
writeMM(counts_matrix, file= "~/t.gondi_dc/data/adata_create/counts.mtx")

# write dimensionality reduction matrix, in this example case pca matrix
write.csv(se.dc.mus@reductions$pca@cell.embeddings, file="~/t.gondi_dc/data/adata_create/pca.csv",quote=F, row.names=F)

# write gene names
write.table(
  data.frame('gene'=rownames(counts_matrix)),file='~/t.gondi_dc/data/adata_create/gene_names.csv',
  quote=F,row.names=F,col.names=F
)

```


```{r, add Average expression of gene sets of interest to the data }

LPS_rows <- which(rownames(se.dc.mus@assays$RNA) %in% gene.set_LPS)
# Get mean expression of genes of interest per cell
mean.exp <- colMeans(x = se.dc.mus@assays$RNA[LPS_rows,], na.rm = TRUE)
# Add mean expression values in 'object@meta.data$gene.set.score'
if (all(names(x = mean.exp) == rownames(x = se.dc.mus@meta.data))) {
  cat("Cell names order match in 'mean.exp' and 'object@meta.data':\n", 
      "adding gene set mean expression values in 'object@meta.data$gene.set.score'")
 se.dc.mus@meta.data$LPS_score <- mean.exp
}
# plot
#png("~/t.gondi_dc/res/GEO_comparison/LPS_score_UMAP.png", width = 15, height = 10, units = "cm", res = 600)
FeaturePlot(object = se.dc.mus, 
            features = "LPS_score", pt.size = 2) +
                              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank()) + 
scale_color_gradientn("mean expression", colors = c("darkblue", "mediumblue", "yellow3", "yellow2", "yellow1", "yellow"))
#dev.off()

```

```{r, visualize gene of interests in UMAP (not included in manuscript)}
###Do the same for PAM specific expression 

PAM_rows <- which(rownames(se.dc.mus@assays$RNA) %in% gene.set_PAM)

# Get mean expression of genes of interest per cell
mean.exp <- colMeans(x = se.dc.mus@assays$RNA[PAM_rows,], na.rm = TRUE)

# Add mean expression values in 'object@meta.data$gene.set.score'
if (all(names(x = mean.exp) == rownames(x = se.dc.mus@meta.data))) {
  cat("Cell names order match in 'mean.exp' and 'object@meta.data':\n", 
      "adding gene set mean expression values in 'object@meta.data$gene.set.score'")
 se.dc.mus@meta.data$PAM_score <- mean.exp
}

# plot
#png("~/t.gondi_dc/res/GEO_comparison/PAM_score_UMAP.png", width = 15, height = 10, units = "cm", res = 600)
FeaturePlot(object = se.dc.mus, 
            features = "PAM_score", pt.size = 2) +
              theme(axis.ticks = element_blank(),
                              axis.text = element_blank(), 
                              axis.title = element_blank(), 
                    legend.position = "right") +
               scale_color_gradientn("mean expression", colors = c("darkblue", "mediumblue", "yellow3", "yellow2","yellow1", "yellow")) 
  
#dev.off()


```


```{r, get the mouse and toxo tables for gr.1 and gr.2 to compare and highlight}


write.table(t.mouse, "~/t.gondi_dc/res/supplementary_data/")
#mouse genes
shared.ldm.ptg <- intersect(unique(t.mouse.ldm$external_gene_name), unique(t.mouse.ptg$external_gene_name))
length(shared.ldm.ptg)
dif.ldm <- setdiff(unique(t.mouse.ldm$external_gene_name), unique(t.mouse.ptg$external_gene_name))
dif.ptg <- setdiff(unique(t.mouse.ptg$external_gene_name), unique(t.mouse.ldm$external_gene_name))


#get the different genes for each condition: 
#remove the shared genes from each dataframe 

int <- which(t.mouse.ldm$external_gene_name %in% shared.ldm.ptg)
ldm.interactors <- t.mouse.ldm[t.mouse.ldm$external_gene_name %in%dif.ldm,]
int <- which(t.mouse.ptg$external_gene_name %in% shared.ldm.ptg)
ptg.interactors <- t.mouse.ptg[t.mouse.ptg$external_gene_name %in% dif.ptg,]
#and get the intersection
 
t.tx.ptg <- get.toxo.table(gr.1, ann = genes.table.tg)
t.tx.ldm <- get.toxo.table(gr.2, ann = genes.table.tg)

shared.ldm.ptg.tx <- intersect(t.tx.ldm$ensembl_gene_id, t.tx.ptg$ensembl_gene_id)
length(shared.ldm.ptg.tx)
dif.ldm.tx <- setdiff(t.tx.ldm$ensembl_gene_id, t.tx.ptg$ensembl_gene_id)
dif.ptg.tx <- setdiff(t.tx.ptg$ensembl_gene_id, t.tx.ldm$ensembl_gene_id)
length(dif.ldm.tx)
length(dif.ptg.tx)
#get the genes names for different and shared toxo genes 
ann <- genes.table.tg
ann$ensembl_gene_id <- gsub("_", ".", ann$ensembl_gene_id)
shared.toxo<- ann[ann$ensembl_gene_id %in% shared.ldm.ptg.tx,]
dif.ldm.tx <- ann[ann$ensembl_gene_id %in% dif.ldm.tx,]
dif.ptg.tx <- ann[ann$ensembl_gene_id %in% dif.ptg.tx,]

dif.ldm.tx <- dif.ldm.tx[dif.ldm.tx$gene_biotype == "protein coding gene" & dif.ldm.tx$description != "hypothetical protein",]
dif.ptg.tx <- dif.ptg.tx[dif.ptg.tx$gene_biotype == "protein coding gene" & dif.ptg.tx$description != "hypothetical protein",]



```

```{r, gprofiler enrichment of interaction genes }


#gprofiler - enrichment

mouse.int.list <- names(clusters(gr.1)$membership[V(gr.1)$name %in% make.names(colnames(TopIntGenes_mouse))])

mus.int <- gost(mouse.int.list, organism = "mmusculus", sources = "GO:BP")

  print(ggplot(head(mus.int$result), aes(reorder(term_name, -log10(p_value)), -log10(p_value))) +
  geom_bar(stat = "identity", color = "black", fill = "darkblue") +
  coord_flip() +
  #scale_fill_gradientn(name = "Enrichment Score",colours =   RColorBrewer::brewer.pal(n = 20, name = "Reds")) + 
  scale_y_continuous() +
  xlab("Biological Process")+
  ylab("Enrichment Score") +
  theme(axis.text.y = element_text(size = 12), axis.title = element_blank(), panel.background = element_blank()))
  
```
We observed that for PTG (type II) strains, there is a decrease of parasite transcripts upon infection progression - thus we would like to investigate whether there are host genes for which expression correlates with the number of Toxoplasma transcripts (in LDM as well as in PTG)

```{r, correlate murine gene expression with number of T. gondii trasncripts}

#Create marices to correlate

#nUMI Toxoxplasma

umiT <- data.frame(nUMI_t = se.dc.mus$toxo_nUMI, condition = se.dc.mus$cell_class, row.names = rownames(se.dc.mus[[]]))

expr.m <- t(as.matrix(se.dc.mus@assays$SCT@counts))


cor.m <- sapply(list(1:ncol(expr.m)), function(gene){
  
  cor(umiT$nUMI_t, expr.m[,gene])
})

cor.t <- sapply(c(1:ncol(expr.m)), function(gene){
  cc <- cor.test(umiT$nUMI_t, expr.m[,gene])
  c <- cc$p.value
  return(c)
})

rownames(cor.m) <- colnames(expr.m)
cor.stat <- as.data.frame(cbind(cor.m, cor.t))
colnames(cor.stat) <- c("correlation_value", "p_value")

cor.stat.s <- subset(cor.stat, cor.stat$p_value < 0.05)

#extract only negative correlations below 0.15
cor.neg <- subset(cor.stat.s, cor.stat.s$correlation_value < -0.15)
cor.neg <- cor.neg[order(cor.neg$correlation_value),]
#extract only positive correlations above 0.15
cor.pos <- subset(cor.stat.s, cor.stat.s$correlation_value > 0.15)
cor.pos <- cor.pos[order(-cor.pos$correlation_value),]


```

As we now have discovered genes which are negatively or positively correlated with the total readCount of Toxoplasma transcripts, we can now determine the expression of these genes as a function of T. gondii specific transcription 

```{r, make graph for mouse-transcription as a function of the Toxo transcription count}

expr_by_dist.tg <- function(object, seed, plotgenes = c(1:length(seed)), ncol = 5, nrow = 4, #thrs = 500,
                            linecol = "forestgreen", ribboncol = "forestgreen") {
  
  
tg <- as.matrix(GetAssayData(object[["SCT"]], slot = "data"))
tg[1:5,1:5]

#get all cells with at least one count for any of the genes of interest
tg <- tg[which(rownames(tg) %in% seed),]
n_toxo <- umiT
tg.goi <- as.data.frame(t(tg))

tg.goi.count <- cbind(tg.goi, n_toxo)
#pb.goi.dist$timepoint <- as.factor(pb.goi.dist$timepoint) #we could add the cluster or the strain here later!

tg.goi.count.pl <- lapply(plotgenes, function(x){

#tg.goi.count.p <- subset(tg.goi.count, tg.goi.count$nUMI_t <= thrs) #no threshold atm

p <- ggplot(tg.goi.count , aes( x= log(nUMI_t), y= tg.goi.count[,x]
                                #, colour = condition
                                )) + 
  #geom_point() +
  #scale_color_manual(values = c("")) +
  geom_smooth(#aes(fill = condition),
              method = "loess", 
              n = 10, span = 0.5) + 
  
  ylab(colnames(tg.goi.count)[x]) + 
  theme_classic()

return(p)
  
})

return(list(out1 = print(head(tg.goi.count)), out2 = ggarrange(plotlist = tg.goi.count.pl, ncol, nrow)))


}

p.pos <- expr_by_dist.tg(se.dc.mus, rownames(cor.pos[1:20,]))
p.pos$out2

p.neg <- expr_by_dist.tg(se.dc.mus, rownames(cor.neg[1:20,]))
p.neg$out2

```

Now we looked over all cells - however, we would like to see differences in these genes only in cells which are infected with T. gondii and also compare the two different strains with each other 


```{r, subset data and split by strain}

expr_by_dist.tg <- function(object, seed, plotgenes = c(1:length(seed)), ncol = 5, nrow = 4,
                            linecol = "forestgreen", ribboncol = "forestgreen") {
  
  
tg <- as.matrix(GetAssayData(object[["SCT"]], slot = "data"))
tg[1:5,1:5]

#get all cells with at least one count for any of the genes of interest
tg <- tg[which(rownames(tg) %in% seed),]
n_toxo <- umiT
#subset umiT
n_toxo <- subset(n_toxo, n_toxo$condition %in% c("PTG_3h", "PTG_12h", "LDM_3h", "LDM_12h"))
tg.goi <- as.data.frame(t(tg))
tg.goi <- subset(tg.goi, rownames(tg.goi) %in% rownames(n_toxo))

tg.goi.count <- cbind(tg.goi, n_toxo)#we could add the cluster or the strain here later!

tg.goi.count.pl <- lapply(plotgenes, function(x){

#tg.goi.count.p <- subset(tg.goi.count, tg.goi.count$nUMI_t <= thrs) #no threshold atm

p <- ggplot(tg.goi.count , aes( x= log(nUMI_t), y= tg.goi.count[,x]
                                , colour = condition
                                )) + 
  #geom_point() +
  #scale_color_manual(values = c("")) +
  geom_smooth(aes(fill = condition),
              method = "loess", 
              n = 10, span = 0.5) + 
  
  ylab(colnames(tg.goi.count)[x]) + 
  theme_classic()

return(p)
  
})

return(list(out1 = print(head(tg.goi.count)), out2 = ggarrange(plotlist = tg.goi.count.pl, ncol, nrow)))


}


p.pos <- expr_by_dist.tg(se.dc.mus, rownames(cor.pos[1:5,]))
p.pos$out2

p.neg <- expr_by_dist.tg(se.dc.mus, rownames(cor.neg[1:5,]))
p.neg$out2

```

It seems like the majority of effects is driven by PTG (type II) at 12h p.i. - Thus I want to try one last attempt to run the correlation analysis per strain and compare the 5 top genes for each of the strains in the plots above. 

```{r, strain specific correlations}


#Create marices to correlate

#nUMI Toxoxplasma

umiT <- data.frame(nUMI_t = se.dc.mus$toxo_nUMI, condition = se.dc.mus$cell_class, row.names = rownames(se.dc.mus[[]]))
umiT.tg <- data.frame(nUMI_t = se.toxo$toxo_nUMI, condition = se.toxo$cell_class, row.names = rownames(se.toxo[[]]))

umiT <- subset(umiT, umiT$condition %in% c("PTG_3h", "PTG_12h", "LDM_3h", "LDM_12h"))
umiT$strain <- ifelse(umiT$condition %in% c("PTG_3h", "PTG_12h"), paste0("typeII"), paste0("typeI"))

#get expression matrix and subset to only one type at the time
expr.m <- as.data.frame(t(as.matrix(se.dc.mus@assays$SCT@counts)))
expr.m.I <- as.data.frame(subset(expr.m, rownames(expr.m) %in% rownames(umiT)[which(umiT$strain == "typeI")]))
expr.m.I <- expr.m.I %>% select(where(~ !is.numeric(.) || sum(.) != 0))
expr.m.II <- as.data.frame(subset(expr.m, rownames(expr.m) %in% rownames(umiT)[which(umiT$condition == "PTG_12h")]))
expr.m.II <- expr.m.II %>% select(where(~ !is.numeric(.) || sum(.) != 0))
expr.m.tg <- as.data.frame(t(as.matrix(se.toxo@assays$SCT@counts)))

umiT_1 <- subset(umiT,umiT$strain == "typeI")
umiT_2 <- subset(umiT,umiT$condition == "PTG_12h")

cor.tgcount.mgene <- function(exp_data, cnt_data ){
  
#  cor.m <- sapply(list(1:ncol(exp_data)), function(gene){
  
#  cor(cnt_data$nUMI_t, data[,gene])
#})

df <- data.frame(label= colnames(exp_data), Estimate="", P.value="")
estimates = ncol(exp_data)
pvalues = ncol(exp_data)
for (i in 1:ncol(exp_data)){
  test <- cor.test(cnt_data[,1], exp_data[,i])
  estimates[i] = test$estimate
  pvalues[i] = test$p.value
}
df$Estimate <- estimates
df$P.value <- pvalues
df

cor.stat.s <- subset(df, df$P.value < 0.05)

#extract only negative correlations below 0.15
cor.neg <- subset(cor.stat.s, cor.stat.s$Estimate < -0.15)
cor.neg <- cor.neg[order(cor.neg$Estimate),]
#extract only positive correlations above 0.15
cor.pos <- subset(cor.stat.s, cor.stat.s$Estimate > 0.15)
cor.pos <- cor.pos[order(-cor.pos$Estimate),]

return(list(cor.neg, cor.pos))
}

typeI <- cor.tgcount.mgene(expr.m.I, umiT_1)
typeII <- cor.tgcount.mgene(expr.m.II, umiT_2)
toxo_cor <- cor.tgcount.mgene(expr.m.tg, umiT.tg)


typeI[2][[1]]$label
typeII[1]



```
Now that we determined strain-wise correlations, we can look at them in our expression-by-Toxopresence plots: 

```{r, strain specific expression correlations}

p.pos.I <- expr_by_dist.tg(se.dc.mus, c(typeI[2][[1]]$label[1:10]))
p.pos.I$out2

p.pos.II <- expr_by_dist.tg(se.dc.mus, c(typeII[2][[1]]$label[1:20]))
p.pos.II$out2



p.neg.I <- expr_by_dist.tg(se.dc.mus, c(typeI[1][[1]]$label[1:10]))
p.neg.I$out2

p.neg.II <- expr_by_dist.tg(se.dc.mus, c(typeII[1][[1]]$label[1:10]))
p.neg.II$out2

```

For the Toxoplasma genes which are positively or negatively expressed we can now see if there is enrichment for specific pathways that might indicate the state of the parasite e.g. apoptosis or similiar 


```{r,}


pie_toxogo(go.table.int, seed = gsub("-","_",toxo_cor[[2]]$label) , cat = "Computed.GO.Functions", col = "Parent")

p.neg.tg <- expr_by_dist.tg(se.toxo, c(toxo_cor[1][[1]]$label[1:10]))
p.neg.tg$out2

p.pos.tg <- expr_by_dist.tg(se.toxo, c(toxo_cor[2][[1]]$label[1:10]))
p.pos.tg$out2

```


###Toxoplasma analysis

It does not seem to be normalized for the total number of UMIs in each cluster since the % of mouse transcripts is also lower in cluster 2 and 3 (as observed for the percentage of T. gondii umis in these clusters)

Check for the total number of Umis in each cluster

```{r, total number transcripts in each cluster}

#Number of toxoplasma transcripts/cluster
nc <- new.cluster.ids
n <- c("sum.c0", "sum.c1", "sum.c2", "sum.c3", "sum.c4")

cluster.t.p.l <- setNames(sapply(nc, function(x){
subset(se.toxo, subset = seurat_clusters %in% x)
}), nm = n)

cluster.t.p.l <- setNames(lapply(cluster.t.p.l, function(x){
  sum(x@meta.data$nUMI)
}), nm = n) 

cluster.t.p.l <- as.data.frame(unlist(cluster.t.p.l))

names(cluster.t.p.l)[1] <- "percentage"
cluster.t.p.l$cluster <- new.cluster.ids

print(ggplot(cluster.t.p.l, aes(cluster, percentage)) + 
  geom_bar(stat = "identity") + 
  theme_classic() + 
  ylab("nUMI per cluster"))

```

```{r, visualise specific genes of interest on T. gondii analysis}


#Gra15 
FeaturePlot(se.toxo, features = "TGME49-275470")

#Rop16
FeaturePlot(se.toxo, features = "TGME49-262730")

se.toxo$Gra15_expression <- ifelse(se.toxo[["SCT"]]@data["TGME49-275470", ] > 1, "Gra15+", "Gra15-")

se.toxo$Rop16_expression <- ifelse(se.toxo[["SCT"]]@data["TGME49-262730", ] > 1, "Rop16+", "Rop16-")

#One group of LDM toxo as well as PTG toxo seems to be clustering away from the rest of the cells with toxo : It would be interesting to take these cells and visualise these on the umap for the dc clustering as well as the combined umap for toxo and DCs! and then look if they are in a specific cluster!! 


```

Calculate amount of toxo transcripts within each toxo cluster 

```{r, number of toxo transcripts in each cluster}

#Number of toxoplasma transcripts/cluster
nc <- new.cluster.ids
n <- c("sum.c0", "sum.c1", "sum.c2", "sum.c3", "sum.c4")

cluster.t.p.l <- setNames(sapply(nc, function(x){
subset(se.toxo, subset = seurat_clusters %in% x)
}), nm = n)

cluster.t.p.l <- setNames(lapply(cluster.t.p.l, function(x){
  sum(x@meta.data$toxo_nUMI)
}), nm = n) 

cluster.t.p.l <- as.data.frame(unlist(cluster.t.p.l))

names(cluster.t.p.l)[1] <- "percentage"
cluster.t.p.l$cluster <- as.factor(new.cluster.ids)

print(ggplot(cluster.t.p.l, aes(cluster, percentage)) + 
  geom_bar(stat = "identity") + 
  theme_classic() + 
  ylab("nUMI T. gondii per cluster"))

```
compare to the total number of cells in each cluster and the numbers of mouse transcripts per cluster and see if you need to normalize for the total number of cells or if that already happened (double-check)! 

```{r, number of mouse transcripts in each cluster}

#Number of toxoplasma transcripts/cluster
nc <- new.cluster.ids
n <- c("sum.c0", "sum.c1", "sum.c2", "sum.c3", "sum.c4")

cluster.t.p.l <- setNames(sapply(nc, function(x){
subset(se.toxo, subset = seurat_clusters %in% x)
}), nm = n)

cluster.t.p.l <- setNames(lapply(cluster.t.p.l, function(x){
  sum(x@meta.data$mouse_nUMI)
}), nm = n) 

cluster.t.p.l <- as.data.frame(unlist(cluster.t.p.l))

names(cluster.t.p.l)[1] <- "percentage"
cluster.t.p.l$cluster <- as.character(0:3)

print(ggplot(cluster.t.p.l, aes(cluster, percentage)) + 
  geom_bar(stat = "identity") + 
  theme_classic() + 
  ylab("nUMI m. musculus per cluster"))

```

```{r, investigate DEG between toxo in GM-DCs vs GM-Macs, fig.width= 40, fig.height= 20}

se.toxo.cell <- SetIdent(se.toxo.cell, value = "mus_celltypes")

toxo.markers.celltype <- FindAllMarkers(se.toxo.cell, only.pos = T)
head(toxo.markers.celltype)
toxo.markers.celltype <- subset(toxo.markers.celltype, p_val_adj < 0.05)
toxo.markers.celltype <- toxo.markers.celltype[order(-toxo.markers.celltype$avg_log2FC),]
#string-split rownames with .1 behind
mutate(toxo.markers.celltype, gene = sapply(strsplit(rownames(toxo.markers.celltype), split ='.', fixed=TRUE),function(x) (x[1])))
  
toxo.markers.DC <- subset(toxo.markers.celltype, cluster == "GM-DC")
toxo.markers.MAC <- subset(toxo.markers.celltype, cluster == "GM-Mac")
nrow(toxo.markers.MAC)
nrow(toxo.markers.DC)

FeaturePlot(se.toxo, features = rownames(toxo.markers.MAC))

png("~/t.gondi_dc/res/figures/fig5/tx_macs-enriched.png", width = 30, height= 20, res = 600, units = "cm")
VlnPlot(se.toxo.cell, features = rownames(toxo.markers.MAC)[c(1,14,14,38)], group.by = "mus_celltypes",  ncol = 2, cols = c("turquoise4", "mediumorchid4"))
dev.off()

png("~/t.gondi_dc/res/figures/fig5/tx_dc-enriched.png", width = 30, height= 10, res = 600, units = "cm")
VlnPlot(se.toxo.cell, features = toxo.markers.DC$gene , group.by = "mus_celltypes", ncol = 2, cols = c("turquoise4", "mediumorchid4"))
dev.off()

```


```{r, unused functions}


##average heatmap function 

#data refers to a sc expression matrix you want to perform the analysis on 
#meta refers to the data you want to group your data by (e.g. clusters, celltypes, etc.)
#cond refers to the conditions you would like to display in the heatmap (numbers 1:16) (default is all conditions)
#seed refers to the features you would like to include (charactervector of genes to display)






## Look at expression of cdks/cyclins across clusters /cell_class

#cluster = vector with cell and clusterannotation/celltype annotation or annotation of interest
#gene = vector with expression values for a gene of interest in the data
#dataset = which dataset to use (dc, dc+mac,mac)
#cols = colors used for the clusters 

CyCDK_expr <- function(object, clust, gene, cols){
  #cyclin B
  gg <- data.frame(cluster = clust, val = object@assays$SCT@data[gene,])
  # x should be the continuous variable and y the categorical group variable
  # You can color the ridges using any variable you want
  gg <- ggplot(gg, aes(x = val, y = cluster, fill = cluster)) +
    geom_density_ridges(alpha = 0.8) +# Set transparency
    scale_fill_manual(values = cols) +
    theme_ridges() +
    labs(x = "Expression value", y = "Cluster identity", fill = "Cluster identity") + 
    ggtitle(gene)
  return(gg)
}


golysis <- function(data, cluster){
  dat <- list(data[data[,2] == cluster,][,1])
  res <- gost(query = dat, organism = "mmusculus", sources = "GO:BP", significant = F)
  
  g <- ggplot(head(res$result, 10), aes(reorder(term_name, -log10(p_value)), -log10(p_value))) +
    geom_bar(stat = "identity", color = "black", fill = "darkblue") +
    coord_flip() +
    #scale_fill_gradientn(name = "Enrichment Score",colours =   RColorBrewer::brewer.pal(n = 20, name = "Reds")) + 
    scale_y_continuous() +
    xlab("Biological Process")+
    ylab("Enrichment Score") +
    theme(axis.text.y = element_text(size = 12), axis.title = element_blank(), panel.background = element_blank())
  return(g)
}

```

CRISPR comparative analysis 


```{r, check which genes in the INF survival genes are GRAs and ROPS}

rop_intersect <- intersect(rop_genes$ensembl_gene_id,crisp_INF_vs_naive$ME49_ID)
#3 rophtry genes 
gra_intersect <- intersect(gra_genes$ensembl_gene_id, crisp_INF_vs_naive$ME49_ID)
#0gra genes 

```

```{r, make correlation tables for gra/rop clusters (virulent/non-virulant invasive or whatever and remaining toxo genes)}

M_toxo <- as.matrix(se.toxo@assays$SCT@counts)
ncol(M_toxo)

rop_gra_genes$ensembl_gene_id <- gsub( "_","-",rop_gra_genes$ensembl_gene_id)

#get first correlation cluster (manually)
virulent.c <- c("TGME49−239600 :: ROP23","TGME49−270920 :: ROP32","TGME49−258800 :: ROP31", "TGME49−295125 :: ROP4","TGME49−275850 :: GRA12","TGME49−243730 :: ROP9", "TGME49−242240 :: ROP19A", "TGME49−262730 :: ROP16", "TGME49−315490 :: ROP10","TGME49−211260 :: ROP26", "TGME49−295110 :: ROP7", "TGME49−315220 :: ROP14","TGME49−258230 :: ROP20","TGME49−262050 :: ROP39","TGME49−205250 :: ROP18", "TGME49−308090 :: ROP5", "TGME49−203990 :: ROP12", "TGME49−227810 :: ROP11", "TGME49−291960 :: ROP40","TGME49−294560 :: ROP37", "TGME49−309590 :: ROP1", "TGME49−215775 :: ROP8", "TGME49−211290 :: ROP15", "TGME49−252360 :: ROP24","TGME49−258580 :: ROP17", "TGME49−258660 :: ROP6", "TGME49−312270 :: ROP13")

which(rop_gra_genes$external_gene_name %in% virulent.c)
virulent.c <- sapply(virulent.c,
               function(x) {strsplit(x,' :: ')[[1]][2]})
names(virulent.c) <- NULL
virulent.c <- rop_gra_genes[rop_gra_genes$external_gene_name %in% virulent.c,]
virulent.c <- virulent.c$ensembl_gene_id
Rops_m <- M_toxo[rownames(M_toxo) %in% virulent.c,]
Rops_m[1:5,1:5]


##get second cluster 
interaction.c <- c("TGME49−266100 :: ROP41", "TGME49−268900 :: GRA10", "TGME49−207700 :: ROP22", "TGME49−242110 :: ROP38", "TGME49−251540 :: GRA9", "TGME49−254720 :: GRA8", "TGME49−275470 :: GRA15", "TGME49−263220 :: ROP21", "TGME49−239740 :: GRA14", "TGME49−201130 :: ROP33", "TGME49−310780 :: GRA4", "TGME49−288650 :: GRA12", "TGME49−304740 :: ROP35", "TGME49−286450 :: GRA5","TGME49−203310 :: GRA7", "TGME49−270250 :: GRA1", "TGME49−227620 :: GRA2", "TGME49−202780 :: ROP25", "TGME49−227280 :: GRA3", "TGME49−227010 :: ROP30", "TGME49−313330 :: ROP27", "TGME49−275440 :: GRA6")

interaction.c <- sapply(interaction.c,function(x) {strsplit(x,' :: ')[[1]][2]})
names(interaction.c) <- NULL
interaction.c <- rop_gra_genes[rop_gra_genes$external_gene_name %in% interaction.c,]
interaction.c <- interaction.c$ensembl_gene_id
GRA_m <- M_toxo[rownames(M_toxo) %in% interaction.c,]
GRA_m[1:5,1:5]

```

```{r, go-term plotting (gprofiler), fig.width= 20, fig.height= 12}

#Fix the data to display it in treeplots 
golysis.tree <- function(data, source, nterms, significant){
  dat <- list(data)
  res <- gost(query = dat, organism = "mmusculus", sources = source , significant = significant)
  res <- res$result
  res$term_go_id <- paste0(res$term_name,"::",res$term_id)
  #res <- subset(res,res$significant == T)
  g <- ggplot(head(res, nterms), aes(area= -log10(p_value), fill= term_name, label = term_go_id)) +
    geom_treemap() + 
    geom_treemap_text(colour = "white",
                    place = "centre",
                    size = 8,
                    grow = T)+
    scale_fill_manual(values = c("#7b2cbf", "#005f73", "#0a9396","#006ac7","#e85d04","#ca6702","#bb3e03","#d00000","#ae2012", "#9b2226","#b5179e", "#ffafcc", "#81b29a", "#007200", "#cbf3f0", "#40916c") ) +
    theme(legend.position = "none")
  return(g)
}


pdf("~/t.gondi_dc/res/figures/fig4/GO_treemap_virulence.pdf",width = 25, height = 12.5)
golysis.tree(tab.n[,1],source = "GO:BP", nterms = 10)
dev.off()

golysis(tab.n, 3)



```

```{r, alternative go enrichmnent and visualization}

  library(tidyverse)
  library(rrvgo)
  library(treemapify)
  library(clusterProfiler)
  library(enrichplot)


#write a function to run the analysis 


BP.data <-  enrichGO(gene          = tab9$external_gene_name,nterms,
                       keyType       = "SYMBOL",
                       ont           = "BP",  
                       pAdjustMethod = "holm", # Here you can adjust the pValue adjustment method - options are : "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"
                       pvalueCutoff  = 0.01,
                       OrgDb = org.Mm.eg.db)


#Convert results to dataframe
dataBP <- as.data.frame(BP.data@result)
dataBP$Type <- "BP"


p.value <- 0.05

#####Data prep BP#####
  data.bp <- 
    dataBP %>%
    dplyr::filter(Type == "BP") %>%                   #Filter on the set p value of (all values)
    mutate(GO = str_extract(ID, "GO:\\d*")) %>%      #Make new column GO with data from the ID column if it has a format of "GO: + any number of digits" (all values)
    dplyr::select(GO, P = p.adjust)                    #select only some specific columns and change the name of column p.adjust to P.
  
  plot_mat.bp <- 
    calculateSimMatrix(data.bp$GO,                     #Make a similarity matrix with a vector containing GOterms determining similarity based on GO term db(?)
                       orgdb="org.Mm.eg.db",
                       ont="BP",       
                       method="Rel") 
  plot_mat2.bp <- 
    reduceSimMatrix(plot_mat.bp,                       #reduces (groups) GO terms together based on their similarity in plot_mat if above a threshold value.
                    setNames(-log10(data.bp$P), data.bp$GO), #Attach the log value of the P values to the similarity matrix, this works as weight somehow (probably in the plot?)
                    threshold=0.7,                     #Just the standard number here: similarity threshold (0-1). Some guidance: Large (allowed similarity=0.9), Medium (0.7), Small (0.5), Tiny (0.4) Defaults to Medium (0.7)
                    orgdb="org.Mm.eg.db") 

  df.BP <- plot_mat2.bp %>%                            #turn into tibble and add a column to show what GOterm type the df contains.
    as_data_frame() %>%
    mutate(Go.col = "BP")
  
df.BP <- df.BP[order(df.BP$size, decreasing = T), ]
#Plot
  #pdf("~/t.gondi_dc/res/GO_treemap_Rop.pdf")
  ggplot(df.BP, aes(area= size, fill= term, label = paste(go,"::",term))) +
    geom_treemap() + 
    geom_treemap_text(colour = "white",
                    place = "centre",
                    size = 8,
                    grow = F)# +
    #scale_fill_manual(values = c("#001219",
      #"#005f73", "#0a9396","#0096c7","#e85d04","#ca6702","#bb3e03","#d00000","#ae2012", "#9b2226", "maroon4","darkgreen","blue4") ) +
   # theme(legend.position = "none")
  #dev.off()
  
  
```

```{r, heatmaps}


#combine rop and gra genes 

rop_gra_genes <- rbind(rop_genes,gra_genes)

t_hmat(hmat_t, rop_genes)
t_hmat(hmat_t, gra_genes)
t_hmat(hmat_c, rop_genes)
t_hmat(hmat_c, gra_genes)
t_hmat(hmat_m, rop_genes)
t_hmat(hmat_m, gra_genes)


rownames(hmat_rop) <- paste(rop_genes$ensembl_gene_id, "::",rop_genes$description)
rownames(hmat_gra) <- paste(gra_genes$ensembl_gene_id, "::",gra_genes$description)

#pdf("~/t.gondi_dc/res/figures/fig4/toxo_clusters_gra_rop.pdf", width = 20, height = 5)
t_hmat(hmat_t, unique(unique(rbind(tab1.tg, tab9.tg))))
t_hmat(hmat_m, unique(tx.markers[tx.markers$cluster=="4",]$gene))


rop_gra_genes$ensembl_gene_id <- gsub( "_","-",rop_gra_genes$ensembl_gene_id)
mat1 <- subset(hmat_t, rownames(hmat_t) %in% rop_gra_genes$ensembl_gene_id)
rownames(mat1) <- paste(rop_gra_genes$ensembl_gene_id[-3], "::", rop_gra_genes$external_gene_name[-3])

dist <- dist(mat1)
x <- hclust(dist)
plot(x, labels = NULL, hang = 0.1, check = TRUE,
     axes = TRUE, frame.plot = FALSE, ann = TRUE,
     main = "Cluster Dendrogram",
     sub = NULL, xlab = NULL, ylab = "Height")


```


```{r, check for differences in toxo expression of genes of interest in mouse data}

#add mouse clusters to toxoplasma object 

mouse_clusters <- data.frame(mouse_cluster = se.dc.mus$seurat_clusters)
mouse_clusters <- subset(mouse_clusters, rownames(mouse_clusters) %in% rownames(se.toxo[[]]))

se.toxo <- AddMetaData(se.toxo, metadata = mouse_clusters, col.name = "mouseclusters")



mus_tox_bar <- function(object, meta, seed, ylab) {
  
dat1 <- na.omit(data.frame(cluster = as.factor(meta), sum_seed = colSums(as.data.frame(object@assays$SCT@counts)[seed,]), sum_total = colSums(as.data.frame(object@assays$RNA@counts))))
#summary statistics
library(dplyr)
print(group_by(dat1, cluster) %>%
  summarise(
    count = n(),
    mean = mean(sum_seed, na.rm = TRUE),
    sd = sd(sum_seed, na.rm = TRUE)
  ))

# Compute the analysis of variance
res.aov <- aov(sum_seed ~ cluster, data = dat1)
# Summary of the analysis
print(summary(res.aov))

tuk <- as.data.frame(TukeyHSD(res.aov)$cluster)
print(tuk$`p adj`)
tuk <- subset(tuk,tuk$`p adj` < 0.05)
print(tuk) 

p1 <- ggplot(dat1, aes(x = cluster, y = sum_seed, fill = cluster)) + 
  geom_boxplot() + 
  scale_fill_manual("cluster",values = c.cols[1:7]) +
  theme(panel.background = element_rect(fill = c("white")), axis.text.y = element_text(size = 10), axis.text.x = element_text(size = 10, angle = 60, hjust = 1), axis.title.x = element_blank(), axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
          axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid'), legend.position = "none") +
  xlab("cluster (m. musculus)") + 
  ylab(ylab)
  
return(p1)
  
}

pdf("~/t.gondi_dc/res/figures/supplementary/box_gene_set_expression_gra.pdf")
mus_tox_bar(se.toxo, se.toxo$mouseclusters, gra_genes$ensembl_gene_id, ylab = "Σ nUMI (GRA)")
dev.off()





```

